<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Class 7 : BEDTools vignette &#8212; Genome Analysis Workshop 0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.6/flatly/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Class 8 : R : Getting Started" href="class-08.html" />
    <link rel="prev" title="Class 6 : Genomic Data Vignette: ChIP-seq" href="class-06.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Genome Analysis Workshop</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Genome Analysis Workshop <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Syllabus.html">Syllabus</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../Classes.html">Class list</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Exercises.html">Exercises</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Problem_Sets.html">Problem Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Miscellaneous.html">Miscellaneous</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul>
<li><a class="reference internal" href="#">Class 7 : BEDTools vignette</a><ul>
<li><a class="reference internal" href="#goals">Goals</a></li>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#annotations">Annotations</a></li>
<li><a class="reference internal" href="#signal">Signal</a></li>
<li><a class="reference internal" href="#slop">Slop</a></li>
<li><a class="reference internal" href="#make-windows">Make windows</a></li>
<li><a class="reference internal" href="#map-signal-to-the-annotations">Map signal to the annotations</a></li>
<li><a class="reference internal" href="#grouping-the-data">Grouping the data</a></li>
<li><a class="reference internal" href="#code">Code</a></li>
<li><a class="reference internal" href="#plotting">Plotting</a></li>
<li><a class="reference internal" href="#exercises">Exercises</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    <div class="col-md-9 content">
      
  <div class="section" id="class-7-bedtools-vignette">
<h1>Class 7 : BEDTools vignette<a class="headerlink" href="#class-7-bedtools-vignette" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#class-7-bedtools-vignette" title="Slides">§</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">2017 Feb 16 Thurs</td>
</tr>
<tr class="field-even field"><th class="field-name">Last updated:</th><td class="field-body">Feb 16, 2017</td>
</tr>
</tbody>
</table>
<div class="section" id="goals">
<h2>Goals<a class="headerlink" href="#goals" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#goals" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Go over BEDTools analysis vignette to show how to combine different
tools</li>
</ol>
<p>There are several more vignettes <a class="reference external" href="https://github.com/arq5x/bedtools-protocols/blob/master/bedtools.md">here.</a>
Looking over these may be helpful to understand some of the questions on
the problem sets.</p>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#overview" title="Slides">§</a></h2>
<p>A common situation in genomic analysis is to examine how signals (e.g.,
coverage in a ChIP-seq experiment) relate to genome annotations (e.g.,
transcription start sites).</p>
<p>This simple process can be broken down into several individual steps:</p>
<ol class="arabic simple">
<li>Obtain genome annotations (we will use TSS)</li>
<li>Obtain signals in bedGraph format (we will use CTCF ChIP-seq)</li>
<li>Use <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/slop.html#slop" title="(in bedtools v2.26.0)"><span class="xref std std-ref">slop</span></a> to create regions to examine [optional]</li>
<li>Break the regions into windows with <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/makewindows.html#makewindows" title="(in bedtools v2.26.0)"><span class="xref std std-ref">makewindows</span></a>.</li>
<li><a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/map.html#map" title="(in bedtools v2.26.0)"><span class="xref std std-ref">map</span></a> data onto the windows.</li>
<li>Calculate a summary statistic for each window with <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/groupby.html#groupby" title="(in bedtools v2.26.0)"><span class="xref std std-ref">groupby</span></a>.</li>
</ol>
</div>
<div class="section" id="annotations">
<h2>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#annotations" title="Slides">§</a></h2>
<p>First we need some annotations. Let&#8217;s get some transcription start sites
(TSSs) from a BED file of gene annotations. For this analysis extract TSSs
from chr22.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ <span class="nv">genes</span><span class="o">=</span><span class="s2">&quot;data-sets/bed/genes.hg19.bed.gz&quot;</span>
$ <span class="nv">tssbed</span><span class="o">=</span><span class="s2">&quot;tss.bed&quot;</span>

<span class="c1"># need to grab the start or end based on the strand field</span>
$ gzcat <span class="nv">$genes</span> <span class="p">|</span> awk <span class="s1">&#39;$6 == &quot;+&quot;&#39;</span> <span class="se">\</span>
  <span class="p">|</span> awk <span class="s1">&#39;BEGIN {OFS=&quot;\t&quot;} {print $1, $2, $2+1, $4}&#39;</span> &gt; <span class="nv">$tssbed</span>
$ gzcat <span class="nv">$genes</span> <span class="p">|</span> awk <span class="s1">&#39;$6 == &quot;-&quot;&#39;</span> <span class="se">\</span>
  <span class="p">|</span> awk <span class="s1">&#39;BEGIN {OFS=&quot;\t&quot;} {print $1, $3, $3+1, $4}&#39;</span> &gt;&gt; <span class="nv">$tssbed</span>

<span class="c1"># extract chr22 intervals and sort bed output</span>
$ awk <span class="s1">&#39;($1 == &quot;chr22&quot;)&#39;</span> <span class="nv">$tssbed</span> <span class="se">\</span>
    <span class="p">|</span> bedtools sort -i - &gt; tmp.bed

<span class="c1">#rename tmp.bed to tss.bed</span>
$ mv tmp.bed <span class="nv">$tssbed</span>
</pre></div>
</div>
</div>
<div class="section" id="signal">
<h2>Signal<a class="headerlink" href="#signal" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#signal" title="Slides">§</a></h2>
<p>The signal we will use is human CTCF ChIP-seq signal. The file is:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">signal</span><span class="o">=</span><span class="s2">&quot;data-sets/bedtools/ctcf.hela.chr22.bg.gz&quot;</span>
</pre></div>
</div>
<p>Another common file format for ChIP/Clip-Seq signals is the <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bigWig.html">bigWig</a> format.</p>
<p>bigWig is a compressed form of bedGraph. You can either convert to bedGraph
and compress:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">bigWigToBedGraph</span> <span class="o">&lt;</span><span class="n">wig</span> <span class="n">file</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">bedgraph</span><span class="o">&gt;</span>
<span class="n">gzip</span> <span class="o">&lt;</span><span class="n">bedgraph</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>Or you can use it in a stream with:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="p">(</span><span class="n">bigWigToBedGraph</span> <span class="o">&lt;</span><span class="n">wigfile</span><span class="o">&gt;</span> <span class="n">stdout</span><span class="p">)</span>
</pre></div>
</div>
<p>Additional Encode data can be downloaded <a class="reference external" href="https://www.encodeproject.org/matrix/?type=Experiment">here</a></p>
</div>
<div class="section" id="slop">
<h2>Slop<a class="headerlink" href="#slop" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#slop" title="Slides">§</a></h2>
<p>Each of the TSS regions we made is 1 base in size. We need to make these
bigger so that we can examine coverage within a larger region. Let&#8217;s make them
2 kb with the <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/slop.html#slop" title="(in bedtools v2.26.0)"><span class="xref std std-ref">slop</span></a> command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ <span class="nv">chromsize</span><span class="o">=</span><span class="s2">&quot;data-sets/bedtools/hg19.genome&quot;</span>
$ <span class="nv">slopbed</span><span class="o">=</span><span class="s2">&quot;tss.slop.2000.bed&quot;</span>
$ bedtools slop -b <span class="m">2000</span> -i <span class="nv">$tssbed</span> -g <span class="nv">$chromsize</span> &gt; <span class="nv">$slopbed</span>
</pre></div>
</div>
<p>Inspect the slop&#8217;d file and make sure you understand how it is different
from the input BED file. Check the new widths of each region using
<code class="docutils literal"><span class="pre">awk</span></code>.</p>
</div>
<div class="section" id="make-windows">
<h2>Make windows<a class="headerlink" href="#make-windows" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#make-windows" title="Slides">§</a></h2>
<p>Now that we have a 2 kb window around each TSS, we can break the regions up
into <cite>windows</cite> and examine signal in each. The idea here is that we would
like to see a peak of signal at the middle window, and it should descrease
as we move away from TSS.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ <span class="nv">windowbed</span><span class="o">=</span><span class="s2">&quot;tss.slop.2000.5bp.windows.bed&quot;</span>

$ bedtools makewindows -b <span class="nv">$slopbed</span> -w <span class="m">5</span> -i srcwinnum <span class="se">\</span>
    <span class="p">|</span> bedtools sort -i - <span class="se">\</span>
    <span class="p">|</span> tr <span class="s2">&quot;_&quot;</span> <span class="s2">&quot;\t&quot;</span> <span class="se">\</span>
    &gt; <span class="nv">$windowbed</span>
</pre></div>
</div>
<p>Note that each window now has a unique number, and the relative positions
of each window of a given number are the same. For example, all the
windows labeled &#8220;1&#8221; are 2000 bases upstream of the original annotated TSS.</p>
<p>What would the signal look like if we didn&#8217;t do this?</p>
</div>
<div class="section" id="map-signal-to-the-annotations">
<h2>Map signal to the annotations<a class="headerlink" href="#map-signal-to-the-annotations" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#map-signal-to-the-annotations" title="Slides">§</a></h2>
<p>Now we can map the signal to each of the windows that we made.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ <span class="nv">signalmap</span><span class="o">=</span><span class="s2">&quot;signal.map.bg&quot;</span>
$ bedtools map -a <span class="nv">$windowbed</span> <span class="se">\</span>
    -b <span class="nv">$signal</span> <span class="se">\</span>
    -c <span class="m">4</span> -o mean -null <span class="m">0</span> <span class="se">\</span>
    &gt; <span class="nv">$signalmap</span>
</pre></div>
</div>
<ul class="simple">
<li>the <code class="docutils literal"><span class="pre">-c</span></code> flag specifies the values to operate on.</li>
<li>the <code class="docutils literal"><span class="pre">-o</span></code> flag specifices the operation to perform (e.g. mean)</li>
<li>What does <code class="docutils literal"><span class="pre">-null</span> <span class="pre">0</span></code> do? Why is it important?</li>
</ul>
<p>Inspect this file and make sure you know what it looks like.</p>
</div>
<div class="section" id="grouping-the-data">
<h2>Grouping the data<a class="headerlink" href="#grouping-the-data" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#grouping-the-data" title="Slides">§</a></h2>
<p>Now we can calculate summary statistics on the mapped data with <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/groupby.html#groupby" title="(in bedtools v2.26.0)"><span class="xref std std-ref">groupby</span></a>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="c1"># Note that the mapped data has to be sorted by window number</span>
$ sort -t<span class="s1">$&#39;\t&#39;</span> -k5,5n <span class="nv">$signalmap</span> <span class="se">\</span>
    <span class="p">|</span> bedtools groupby <span class="se">\</span>
        -i - <span class="se">\</span>
        -g <span class="m">5</span> -c <span class="m">6</span> -o sum <span class="se">\</span>
        &gt; output.tab
</pre></div>
</div>
<ul class="simple">
<li>the <code class="docutils literal"><span class="pre">-g</span></code> flag specifies the column to group on</li>
<li>the <code class="docutils literal"><span class="pre">-c</span></code> flag specifies the values to aggregate.</li>
<li>the <code class="docutils literal"><span class="pre">-o</span></code> flag specifices the operation to perform (e.g. sum)</li>
</ul>
<p>Inspect the output so you know what it looks like.</p>
</div>
<div class="section" id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#code" title="Slides">§</a></h2>
<div class="highlight-bash"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env bash</span>

<span class="nv">genes</span><span class="o">=</span><span class="s2">&quot;data-sets/bed/genes.hg19.bed.gz&quot;</span>
<span class="nv">tssbed</span><span class="o">=</span><span class="s2">&quot;tss.bed&quot;</span>

<span class="c1"># make TSS annotations</span>

gzcat <span class="nv">$genes</span> <span class="p">|</span> awk <span class="s1">&#39;$6 == &quot;+&quot;&#39;</span> <span class="se">\</span>
 <span class="p">|</span> awk <span class="s1">&#39;BEGIN {OFS=&quot;\t&quot;} {print $1,$2,$2+1,$4}&#39;</span> &gt; <span class="nv">$tssbed</span>
gzcat <span class="nv">$genes</span> <span class="p">|</span> awk <span class="s1">&#39;$6 == &quot;-&quot;&#39;</span> <span class="se">\</span>
 <span class="p">|</span> awk <span class="s1">&#39;BEGIN {OFS=&quot;\t&quot;} {print $1,$3,$3+1,$4}&#39;</span> &gt;&gt; <span class="nv">$tssbed</span>

<span class="c1"># extract chr22 intervals and sort bed output</span>
awk <span class="s1">&#39;($1 == &quot;chr22&quot;)&#39;</span> <span class="nv">$tssbed</span> <span class="se">\</span>
  <span class="p">|</span> bedtools sort -i - &gt; tmp.bed
mv tmp.bed <span class="nv">$tssbed</span>

<span class="nv">signal</span><span class="o">=</span><span class="s2">&quot;data-sets/bedtools/ctcf.hela.chr22.bg.gz&quot;</span>

<span class="nv">chromsize</span><span class="o">=</span><span class="s2">&quot;data-sets/bedtools/hg19.genome&quot;</span>
<span class="nv">slopbed</span><span class="o">=</span><span class="s2">&quot;tss.slop.2000.bed&quot;</span>

<span class="c1"># make slop</span>
bedtools slop -b <span class="m">2000</span> -i <span class="nv">$tssbed</span> -g <span class="nv">$chromsize</span> &gt; <span class="nv">$slopbed</span>

<span class="c1"># make windows</span>
<span class="nv">windowbed</span><span class="o">=</span><span class="s2">&quot;tss.slop.2000.5bp.windows.bed&quot;</span>
bedtools makewindows -b <span class="nv">$slopbed</span> -w <span class="m">5</span> -i srcwinnum <span class="se">\</span>
    <span class="p">|</span> sort -k1,1 -k2,2n <span class="se">\</span>
    <span class="p">|</span> tr <span class="s2">&quot;_&quot;</span> <span class="s2">&quot;\t&quot;</span> <span class="se">\</span>
    &gt; <span class="nv">$windowbed</span>

<span class="c1"># map data</span>
<span class="nv">signalmap</span><span class="o">=</span><span class="s2">&quot;signal.map.bg&quot;</span>
bedtools map -a <span class="nv">$windowbed</span> <span class="se">\</span>
    -b <span class="nv">$signal</span> <span class="se">\</span>
    -c <span class="m">4</span> -o mean -null <span class="m">0</span> <span class="se">\</span>
    &gt; <span class="nv">$signalmap</span> 

<span class="c1"># group data</span>
sort -t<span class="s1">$&#39;\t&#39;</span> -k5,5n <span class="nv">$signalmap</span> <span class="se">\</span>
    <span class="p">|</span> bedtools groupby <span class="se">\</span>
        -i - <span class="se">\</span>
        -g <span class="m">5</span> -c <span class="m">6</span> -o sum <span class="se">\</span>
        &gt; output.tab 

<span class="c1"># inspect output.tab to see how to plot the data</span>
</pre></div>
</td></tr></table></div>
</div>
<div class="section" id="plotting">
<h2>Plotting<a class="headerlink" href="#plotting" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#plotting" title="Slides">§</a></h2>
<p>Now we&#8217;ll make a plot with R. Navigate to:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">amc</span><span class="o">-</span><span class="n">tesla</span><span class="o">.</span><span class="n">ucdenver</span><span class="o">.</span><span class="n">pvt</span><span class="o">/</span><span class="n">rstudio</span>
</pre></div>
</div>
<p>and login with your tesla credentials. You should see R Studio open up.
Navigate to the directory where you did you work with the <cite>Files</cite> menu on
the lower right, and then set the working directory with <cite>More</cite> &#8211;&gt; <cite>Set
as Working Directory</cite>.</p>
<p>Now make a plot!</p>
<div class="highlight-r"><div class="highlight"><pre><span></span><span class="o">&gt;</span> install.packages<span class="p">(</span>ggplot2<span class="p">)</span>
<span class="o">&gt;</span> <span class="kn">library</span><span class="p">(</span>ggplot2<span class="p">)</span>
<span class="o">&gt;</span> col.names <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&#39;window&#39;</span><span class="p">,</span><span class="s">&#39;signal&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> df <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&#39;output.tab&#39;</span><span class="p">,</span> col.names<span class="o">=</span>col.names<span class="p">)</span>

<span class="c1"># coerce to numbers</span>
<span class="o">&gt;</span> df<span class="o">$</span>signal <span class="o">&lt;-</span> <span class="kp">as.double</span><span class="p">(</span>df<span class="o">$</span>signal<span class="p">)</span>
<span class="o">&gt;</span> df<span class="o">$</span>window <span class="o">&lt;-</span> <span class="kp">seq</span><span class="p">(</span><span class="m">-2000</span><span class="p">,</span> <span class="m">2000</span><span class="p">,</span> <span class="m">5</span><span class="p">)</span>
<span class="o">&gt;</span> qplot<span class="p">(</span>window<span class="p">,</span> signal<span class="p">,</span> data <span class="o">=</span> df<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="exercises">
<h2>Exercises<a class="headerlink" href="#exercises" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-07.html#exercises" title="Slides">§</a></h2>
<p>Try to generalize this approach to other types of data. How would you:</p>
<ol class="arabic simple">
<li>Look at CTCF distribution near transcription termination sites
(TTSs)? Write this program and run it. (easy)</li>
<li>Analyze the distribution of H3K36me3 ChIP-seq signal relative to coding
exons?  Download some H3K36me3 data and <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/bed12tobed6.html#bed12tobed6" title="(in bedtools v2.26.0)"><span class="xref std std-ref">bed12tobed6</span></a>.</li>
<li>Analyze the distribution of H3K36me3 ChIP-seq relative to introns?
Use the <code class="docutils literal"><span class="pre">-n</span></code> option in <a class="reference external" href="http://bedtools.readthedocs.io/en/latest/content/tools/makewindows.html#makewindows" title="(in bedtools v2.26.0)"><span class="xref std std-ref">makewindows</span></a>.</li>
<li>Determine levels of evolutionary conservation surrounding different
types of non-coding variants? (medium). Use dbSnp and phyloP scores.</li>
</ol>
</div>
</div>


    </div>
      
  </div>
</div>
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>

  </body>
</html>