<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>BEDTools vignette &mdash; Genome Analysis Workshop 0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/flatly/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Problem Sets" href="../Problem_Sets.html" />
    <link rel="prev" title="Class 6 : Genomic Data Vignette: ChIP-seq" href="class-6.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Genome Analysis Workshop</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="http://hesselberthlab.github.io">Hesselberth lab</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Genome Analysis Workshop <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Syllabus.html">Syllabus</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../Classes.html">Class list</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Problem_Sets.html">Problem Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Miscellaneous.html">Miscellaneous</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul>
<li><a class="reference internal" href="#">BEDTools vignette</a><ul>
<li><a class="reference internal" href="#goals">Goals</a></li>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#annotations">Annotations</a></li>
<li><a class="reference internal" href="#signal">Signal</a></li>
<li><a class="reference internal" href="#slop">Slop</a></li>
<li><a class="reference internal" href="#make-windows">Make windows</a></li>
<li><a class="reference internal" href="#map-signal-to-the-annotations">Map signal to the annotations</a></li>
<li><a class="reference internal" href="#grouping-the-data">Grouping the data</a></li>
<li><a class="reference internal" href="#plotting">Plotting</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    <div class="col-md-9">
      
  <div class="section" id="bedtools-vignette">
<h1>BEDTools vignette<a class="headerlink" href="#bedtools-vignette" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#bedtools-vignette" title="Slides">§</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">T Feb 17</td>
</tr>
</tbody>
</table>
<div class="section" id="goals">
<h2>Goals<a class="headerlink" href="#goals" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#goals" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Go over BEDTools analysis vignette to show how to combine different
tools</li>
</ol>
<p>There are several vignettes <cite>here.
&lt;https://github.com/arq5x/bedtools-protocols/blob/master/bedtools.md&gt;_</cite></p>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#overview" title="Slides">§</a></h2>
<p>A common situation in genomic analysis is to examine how signals (e.g.,
coverage in a ChIP-seq experiment) relate to genome annotations (e.g.,
transcription start sites).</p>
<p>This simple process can be broken down into several individual steps:</p>
<ol class="arabic simple">
<li>Obtain genome annotations (we will use TSS)</li>
<li>Obtain signals in bedGraph format (we will use Pol II ChIP-seq)</li>
<li>Use <cite>slop &lt;bedtools:slop&gt;</cite> to create regions to examine [optional]</li>
<li>Break the regions into windows with <cite>makewindows
&lt;bedtools:makewindows&gt;</cite>.</li>
<li><cite>map &lt;bedtools:map&gt;</cite> data onto the windows.</li>
<li>Calculate a summary statistic for each window with <cite>groupby
&lt;bedtools:groupby&gt;</cite>.</li>
</ol>
</div>
<div class="section" id="annotations">
<h2>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#annotations" title="Slides">§</a></h2>
<p>First we need some annotations. Let&#8217;s get some transcription start sites
(TSSs) from a BED file of gene annotations.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ genes</span><span class="o">=</span>/vol1/opt/data/refGene.bed.gz
<span class="nv">$ tssbed</span><span class="o">=</span>tss.bed

<span class="c"># need to grab the start or end based on the strand field</span>
<span class="nv">$ </span>zcat <span class="nv">$genes</span> <span class="p">|</span> awk <span class="s1">&#39;$6 == &quot;+&quot;&#39;</span> <span class="p">|</span> awk <span class="s1">&#39;{print $1,$2,$2+1}&#39;</span> &gt; <span class="nv">$tssbed</span>
<span class="nv">$ </span>zcat <span class="nv">$genes</span> <span class="p">|</span> awk <span class="s1">&#39;$6 == &quot;-&quot;&#39;</span> <span class="p">|</span> awk <span class="s1">&#39;{print $1,$3,$3+1}&#39;</span> &gt; <span class="nv">$tssbed</span>

<span class="c"># we made a bed file, so now we sort it.</span>
<span class="nv">$ </span>bedSort <span class="nv">$tssbed</span> <span class="nv">$tssbed</span>
<span class="nv">$ </span>gzip <span class="nv">$tssbed</span>
</pre></div>
</div>
</div>
<div class="section" id="signal">
<h2>Signal<a class="headerlink" href="#signal" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#signal" title="Slides">§</a></h2>
<p>The signal we will use is human Pol II ChIP-seq signal. The file is:</p>
<div class="highlight-python"><div class="highlight"><pre>signal=/vol1/opt/data/encode/wgEncodeBroadHistoneHelas3Pol2bStdSig.bigWig
</pre></div>
</div>
<p>bigWig is a compressed form of bedGraph. You can either convert to bedGraph
and compress:</p>
<div class="highlight-python"><div class="highlight"><pre>bigWigToBedGraph &lt;wig file&gt; &lt;bedgraph&gt;
gzip &lt;bedgraph&gt;
</pre></div>
</div>
<p>Or you can use it in a stream with:</p>
<div class="highlight-python"><div class="highlight"><pre>&lt;(bigWigToBedGraph &lt;wigfile&gt; stdout)
</pre></div>
</div>
</div>
<div class="section" id="slop">
<h2>Slop<a class="headerlink" href="#slop" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#slop" title="Slides">§</a></h2>
<p>Each of the TSS regions we made is 1 base in size. We need to make these
bigger so that we can examine coverage in a larger region. Let&#8217;s make them
2 kb with the <cite>slop &lt;bedtools:slop</cite> command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ chromsize</span><span class="o">=</span>/vol1/opt/data/hg19.chrom.sizes
<span class="nv">$ slopbed</span><span class="o">=</span>tss.slop.2000.bed
<span class="nv">$ </span>bedtools slop -b <span class="m">2000</span> -i <span class="nv">$tssbed</span> -g <span class="nv">$chromsize</span> &gt; <span class="nv">$slopbed</span>
</pre></div>
</div>
<p>romsize=/vol1/opt/data/hg19.chrom.sizesInspect the slop&#8217;d file and make sure you understand how it is different
from the input BED file.</p>
</div>
<div class="section" id="make-windows">
<h2>Make windows<a class="headerlink" href="#make-windows" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#make-windows" title="Slides">§</a></h2>
<p>Now that we have a 2 kb window around each TSS, we can break the regions up
into <cite>windows</cite> and examine signal in each. The idea here is that we would
like to see a peak of signal at the middle window, and it should descrease
as we move away from TSS.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ windowbed</span><span class="o">=</span>tss.slop.2000.5bp.windows.bed

<span class="nv">$ </span>bedtools makewindows -b <span class="nv">$slopbed</span> -w <span class="m">5</span> -i srcwinnum <span class="se">\</span>
    sort -k1,1 -k2,2n <span class="se">\</span>
    tr <span class="s2">&quot;_&quot;</span> <span class="s2">&quot;\t&quot;</span> <span class="se">\</span>
    &gt; <span class="nv">$windowbed</span>
</pre></div>
</div>
<p>What would the signal look like if we didn&#8217;t do this?</p>
</div>
<div class="section" id="map-signal-to-the-annotations">
<h2>Map signal to the annotations<a class="headerlink" href="#map-signal-to-the-annotations" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#map-signal-to-the-annotations" title="Slides">§</a></h2>
<p>Now we can map the signal to each of the windows that we made.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ signalmap</span><span class="o">=</span>signal.map.bg
<span class="nv">$ </span>bedtools map -a <span class="nv">$windowbed</span> <span class="se">\</span>
    -b &lt;<span class="o">(</span>bigWigToBedGraph <span class="nv">$signal</span> stdout<span class="o">)</span> <span class="se">\</span>
    -c <span class="m">4</span> -o mean -null <span class="m">0</span> <span class="se">\</span>
    &gt; <span class="nv">$signalmap</span>
</pre></div>
</div>
<p>Inspect this file and make sure you know what it looks like.</p>
</div>
<div class="section" id="grouping-the-data">
<h2>Grouping the data<a class="headerlink" href="#grouping-the-data" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#grouping-the-data" title="Slides">§</a></h2>
<p>Now we can calculate summary statistics on the mapped data with <cite>groupby
&lt;bedtools:groupby&gt;</cite>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># Note that the mapped data has to be sorted by window number</span>
<span class="nv">$ </span>sort -t<span class="s1">$&#39;\t&#39;</span> -k5,5n <span class="nv">$signalmap</span> <span class="se">\</span>
    bedtools groupby <span class="se">\</span>
        -i - <span class="se">\</span>
        -g <span class="m">5</span> -c <span class="m">6</span> -o sum <span class="se">\</span>
        &gt; output.tab
</pre></div>
</div>
<p>Inspect the output so you know what it looks like.</p>
</div>
<div class="section" id="plotting">
<h2>Plotting<a class="headerlink" href="#plotting" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-7.html#plotting" title="Slides">§</a></h2>
<p>Now we&#8217;ll make a plot with R. Navigate to:</p>
<div class="highlight-python"><div class="highlight"><pre>http://amc-tesla.ucdenver.pvt/rstudio
</pre></div>
</div>
<p>and login with your tesla credentials. You should see R Studio open up.
Navigate to the directory where you did you work with the <cite>Files</cite> menu on
the lower right, and then set the working directory with <cite>More</cite> &#8211;&gt; <cite>Set
as Working Directory</cite>.</p>
<p>Now make a plot!</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> <span class="kn">library</span><span class="p">(</span>ggplot2<span class="p">)</span>
<span class="o">&gt;</span> col.names <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&#39;window&#39;</span><span class="p">,</span><span class="s">&#39;signal&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> df <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&#39;output.tab&#39;</span><span class="p">,</span> col.names<span class="o">=</span>col.names<span class="p">)</span>

<span class="c1"># coerce to numbers</span>
<span class="o">&gt;</span> df<span class="o">$</span>signal <span class="o">&lt;-</span> <span class="kp">as.double</span><span class="p">(</span>df<span class="o">$</span>signal<span class="p">)</span>
<span class="o">&gt;</span> qplot<span class="p">(</span>window<span class="p">,</span> signal<span class="p">,</span> data <span class="o">=</span> df<span class="p">)</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="../_sources/Classes/class-7.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2014, Jay Hesselberth.<br/>
      Last updated on Feb 16, 2015.<br/>
    </p>
  </div>
</footer>
  </body>
</html>