<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Class 9 : R : Manipulation &amp; Plotting &#8212; Genome Analysis Workshop 0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.6/flatly/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Class 15 : Python : Basics" href="class-15.html" />
    <link rel="prev" title="Class 8 : R : Getting Started" href="class-08.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Genome Analysis Workshop</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Genome Analysis Workshop <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Syllabus.html">Syllabus</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../Classes.html">Class list</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Exercises.html">Exercises</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Problem_Sets.html">Problem Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Miscellaneous.html">Miscellaneous</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul>
<li><a class="reference internal" href="#">Class 9 : R : Manipulation &amp; Plotting</a><ul>
<li><a class="reference internal" href="#goals">Goals</a></li>
<li><a class="reference internal" href="#ggplot2-syntax">ggplot2 syntax</a></li>
<li><a class="reference internal" href="#ggplot2-docs">ggplot2 docs</a></li>
<li><a class="reference internal" href="#histograms">Histograms</a></li>
<li><a class="reference internal" href="#exercises">Exercises</a><ul>
<li><a class="reference internal" href="#exercises-with-dplyr">Exercises with dplyr</a></li>
</ul>
</li>
<li><a class="reference internal" href="#combining-with-ggplot">Combining with ggplot</a></li>
<li><a class="reference internal" href="#ggplot-histograms">ggplot histograms</a><ul>
<li><a class="reference internal" href="#id1">Exercises</a></li>
<li><a class="reference internal" href="#practice">Practice</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    <div class="col-md-9 content">
      
  <div class="section" id="class-9-r-manipulation-plotting">
<h1>Class 9 : R : Manipulation &amp; Plotting<a class="headerlink" href="#class-9-r-manipulation-plotting" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#class-9-r-manipulation-plotting" title="Slides">§</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">2017 Feb 23 Thurs</td>
</tr>
<tr class="field-even field"><th class="field-name">Last updated:</th><td class="field-body">Apr 13, 2017</td>
</tr>
</tbody>
</table>
<div class="section" id="goals">
<h2>Goals<a class="headerlink" href="#goals" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#goals" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Begin to make plots with ggplot2 and combine with dplyr</li>
</ol>
</div>
<div class="section" id="ggplot2-syntax">
<h2>ggplot2 syntax<a class="headerlink" href="#ggplot2-syntax" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#ggplot2-syntax" title="Slides">§</a></h2>
<div class="highlight-r"><div class="highlight"><pre><span></span><span class="kn">library</span><span class="p">(</span>tidyverse<span class="p">)</span>
df <span class="o">&lt;-</span> read_tsv<span class="p">(</span><span class="s">&#39;expr-geno-covs.txt&#39;</span><span class="p">)</span>

ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span>

ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
    geom_point<span class="p">()</span>

ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
    geom_point<span class="p">(</span>color<span class="o">=</span><span class="s">&#39;red&#39;</span><span class="p">,</span> size <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">aes()</span></code> stands for <strong>aesthetics</strong>, which specifies the the coordinates,
colors, size, etc from these columns in the data.frame.</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">,</span> color<span class="o">=</span>gender<span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">geom_point()</span></code> means plot these as points, could be <code class="docutils literal"><span class="pre">geom_line()</span></code> or
a number of other <cite>geoms</cite>.</p>
</div>
<div class="section" id="ggplot2-docs">
<h2>ggplot2 docs<a class="headerlink" href="#ggplot2-docs" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#ggplot2-docs" title="Slides">§</a></h2>
<p>The ggplot2 docs are very good: <a class="reference external" href="http://docs.ggplot2.org/current/">http://docs.ggplot2.org/current/</a></p>
<p>Use them find how to change the y-scale on this plot to log10:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>dfx<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
        geom_point<span class="p">()</span>
</pre></div>
</div>
<p>Look at the <cite>geom_point()</cite> documentation and change the color
of the plot above so that males and females are colored differently.</p>
</div>
<div class="section" id="histograms">
<h2>Histograms<a class="headerlink" href="#histograms" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#histograms" title="Slides">§</a></h2>
<p>One of the simplest things to do in R, without ggplot is to look at
a histogram of your data:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>hist<span class="p">(</span>df<span class="o">$</span><span class="kp">expression</span><span class="p">)</span>
<span class="c1"># or</span>
hist<span class="p">(</span><span class="kp">log</span><span class="p">(</span>df<span class="o">$</span><span class="kp">expression</span><span class="p">))</span>
</pre></div>
</div>
<p>You can make these look a lot nicer with ggplot2.</p>
<p><code class="docutils literal"><span class="pre">hist()</span></code> does not work with ggplot, you&#8217;ll have to use the
ggplot2 machinery for that:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span><span class="kp">log10</span><span class="p">(</span><span class="kp">expression</span><span class="p">))</span>
  <span class="o">+</span> geom_histogram<span class="p">()</span>
</pre></div>
</div>
<p>Try to figure out how to overlay a density plot with <code class="docutils literal"><span class="pre">geom_density()</span></code>.</p>
<p>You can also make effective separate visualizations with <code class="docutils literal"><span class="pre">facet_grid()</span></code>
and <code class="docutils literal"><span class="pre">facet_wrap()</span></code>. You need to specify a formula to determine how to
separate:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>gp <span class="o">&lt;-</span> ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span> geom_point<span class="p">()</span>

<span class="c1"># separate by gender</span>
gp <span class="o">+</span> facet_grid<span class="p">(</span><span class="o">~</span> gender<span class="p">)</span>

<span class="c1"># separate by gender and condition</span>
gp <span class="o">+</span> facet_grid<span class="p">(</span>condition <span class="o">~</span> gender<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="exercises">
<h2>Exercises<a class="headerlink" href="#exercises" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#exercises" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Make a histogram using ggplot and separate cases from controls by
changing fill color and symbol type.</li>
<li>Make a histogram using ggplot and separate cases from controls
using <code class="docutils literal"><span class="pre">facet_grid()</span></code>, <code class="docutils literal"><span class="pre">facet_wrap()</span></code>.</li>
</ol>
<div class="section" id="exercises-with-dplyr">
<h3>Exercises with dplyr<a class="headerlink" href="#exercises-with-dplyr" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#exercises-with-dplyr" title="Slides">§</a></h3>
<p>Use the <code class="docutils literal"><span class="pre">expr-geno-covs.txt</span></code> file for the following exercises.</p>
<ol class="arabic simple">
<li>Use <code class="docutils literal"><span class="pre">dplyr</span></code> to calculate the mean age of smokers grouped by gender
and smoking status. Plot the result.</li>
<li>Make a plot of age by expression faceted by genotype. Fit a linear
model through these curves (use geom_smooth) on the plot.</li>
</ol>
</div>
</div>
<div class="section" id="combining-with-ggplot">
<h2>Combining with ggplot<a class="headerlink" href="#combining-with-ggplot" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#combining-with-ggplot" title="Slides">§</a></h2>
<p>We had mean expression by condition and genotype as:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>df <span class="o">%&gt;%</span> group_by<span class="p">(</span>condition<span class="p">,</span> genotype<span class="p">)</span> \
    <span class="o">%&gt;%</span> summarize<span class="p">(</span>count <span class="o">=</span> n<span class="p">(),</span>
                  expr.mean <span class="o">=</span> <span class="kp">mean</span><span class="p">(</span><span class="kp">expression</span><span class="p">))</span>
</pre></div>
</div>
<p>We can continue the thought by adding a plot with the pipe:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span><span class="o">%&gt;%</span> ggplot<span class="p">(</span>aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> \
    <span class="o">+</span> geom_histogram<span class="p">(</span>stat<span class="o">=</span><span class="s">&#39;identity&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>how can we change the color of all the bars to &#8216;red&#8217;? [Hint, it&#8217;s not
<strong>color</strong> =&#8217;red&#8217;]</p>
</div>
<div class="section" id="ggplot-histograms">
<h2>ggplot histograms<a class="headerlink" href="#ggplot-histograms" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#ggplot-histograms" title="Slides">§</a></h2>
<p>Since <cite>expr-geno-covs.txt</cite> is already in long format, we can use it directly in
ggplot:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
       geom_histogram<span class="p">()</span> <span class="o">+</span>
       scale_x_log10<span class="p">()</span>
</pre></div>
</div>
<div class="section" id="id1">
<h3>Exercises<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#id1" title="Slides">§</a></h3>
<ol class="arabic simple">
<li>Adjust this:</li>
</ol>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
       geom_histogram<span class="p">()</span> <span class="o">+</span>
       scale_x_log10<span class="p">()</span>
</pre></div>
</div>
<ul class="simple">
<li>to color by genotype</li>
<li>and to split plots (facet_wrap) by condition (case/control)</li>
<li>to color by age &gt; 60 vs. &lt;= 60 (use row selection stuff from start of class to
make a new column named, e.g. <cite>is_old</cite>)</li>
</ul>
<ol class="arabic simple">
<li>Figure out how to move overlapping points so categorical data is
viewable (hint: look at geom_jitter() or the <cite>position</cite> argument to
geom_point())</li>
<li>Load a BED file (e.g. <code class="docutils literal"><span class="pre">lamina.bed</span></code>) and calculate the mean length of
regions on each chromosome in the BED file with dplyr.  Plot the result as
a bar plot with ggplot2.</li>
</ol>
</div>
<div class="section" id="practice">
<h3>Practice<a class="headerlink" href="#practice" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-09.html#practice" title="Slides">§</a></h3>
<p>Start the Datacamp dplyr and ggplot2 courses if you haven&#8217;t already.</p>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>

  </body>
</html>