<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Week 3 : Working with genomic data &mdash; Genome Analysis Workshop 0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/flatly/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Problem Sets" href="../Problem_Sets.html" />
    <link rel="prev" title="Class 4 : awk" href="class-4.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Genome Analysis Workshop</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="http://hesselberthlab.github.io">Hesselberth lab</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Genome Analysis Workshop <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Syllabus.html">Syllabus</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../Classes.html">Class list</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Problem_Sets.html">Problem Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Miscellaneous.html">Miscellaneous</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul>
<li><a class="reference internal" href="#">Week 3 : Working with genomic data</a><ul>
<li><a class="reference internal" href="#goals">Goals</a></li>
<li><a class="reference internal" href="#grep">grep</a></li>
<li><a class="reference internal" href="#cluster-access">Cluster access</a></li>
<li><a class="reference internal" href="#cluster-etiquette">Cluster etiquette</a></li>
<li><a class="reference internal" href="#example-commands-on-the-cluster">Example commands on the cluster</a></li>
<li><a class="reference internal" href="#the-queueing-system">The queueing system</a></li>
<li><a class="reference internal" href="#killing-jobs">Killing jobs</a></li>
<li><a class="reference internal" href="#other-cluster-specific-commands">Other cluster-specific commands</a></li>
<li><a class="reference internal" href="#questions">Questions</a></li>
<li><a class="reference internal" href="#more-exercises">More exercises</a></li>
</ul>
</li>
<li><a class="reference internal" href="#bedtools">BEDTools</a><ul>
<li><a class="reference internal" href="#id1">Goals</a></li>
<li><a class="reference internal" href="#bedtools-overview">BEDTools Overview</a></li>
<li><a class="reference internal" href="#bedtools-utility">BEDTools Utility</a></li>
<li><a class="reference internal" href="#bedtools-commands">BEDTools Commands</a></li>
<li><a class="reference internal" href="#bedtools-documentation">BEDTools Documentation</a></li>
<li><a class="reference internal" href="#bedtools-intersect">BEDTools intersect</a></li>
<li><a class="reference internal" href="#example-files">Example Files</a></li>
<li><a class="reference internal" href="#intersect">intersect</a></li>
<li><a class="reference internal" href="#intersect-wa">intersect -wa</a></li>
<li><a class="reference internal" href="#intersect-wo">intersect -wo</a></li>
<li><a class="reference internal" href="#intersect-exercise">intersect exercise</a></li>
<li><a class="reference internal" href="#intersect-c">intersect -c</a></li>
<li><a class="reference internal" href="#intersect-v">intersect -v</a></li>
<li><a class="reference internal" href="#intersect-summary">Intersect Summary</a></li>
<li><a class="reference internal" href="#exercises-or-other-tools">Exercises (Or Other Tools)</a></li>
<li><a class="reference internal" href="#other-reading">Other Reading</a></li>
<li><a class="reference internal" href="#intersect-bam">Intersect Bam</a></li>
<li><a class="reference internal" href="#intersect-strand">Intersect Strand</a></li>
<li><a class="reference internal" href="#closest">Closest</a></li>
<li><a class="reference internal" href="#map">Map</a></li>
<li><a class="reference internal" href="#sorted">Sorted</a></li>
<li><a class="reference internal" href="#genomecov">Genomecov</a></li>
<li><a class="reference internal" href="#encode">ENCODE</a></li>
<li><a class="reference internal" href="#encode-project-cell-lines">ENCODE Project Cell Lines</a></li>
<li><a class="reference internal" href="#experiments">Experiments</a></li>
<li><a class="reference internal" href="#common-file-formats">Common File Formats</a></li>
<li><a class="reference internal" href="#references">References</a></li>
<li><a class="reference internal" href="#how-to-access-encode-data">How to Access ENCODE Data</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    <div class="col-md-9">
      
  <div class="section" id="week-3-working-with-genomic-data">
<h1>Week 3 : Working with genomic data<a class="headerlink" href="#week-3-working-with-genomic-data" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#week-3-working-with-genomic-data" title="Slides">§</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">Tues 10 Feb 2015</td>
</tr>
<tr class="field-even field"><th class="field-name">Class date:</th><td class="field-body">Thurs 12 Feb 2015</td>
</tr>
</tbody>
</table>
<div class="section" id="goals">
<h2>Goals<a class="headerlink" href="#goals" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#goals" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Learn to run scripts on the cluster via the queuing system</li>
<li>Learn about genomic data types and where to get data</li>
<li>Start to use <tt class="docutils literal"><span class="pre">bedtools</span></tt> to analyze genomic data</li>
</ol>
</div>
<div class="section" id="grep">
<h2>grep<a class="headerlink" href="#grep" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#grep" title="Slides">§</a></h2>
<p>Use <a class="reference external" href="http://linux.die.net/man/1/grep">grep(1)</a> to identify lines in a file that match a specified pattern.</p>
<p>To find any instance of <em>chr5</em> in the lamina.bed file</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># grep [pattern] [filename]</span>
<span class="nv">$ </span>grep chr5 /vol1/opt/data/lamina.bed <span class="p">|</span> head
</pre></div>
</div>
<p>To find all lines that start with a number sign:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># The caret (^) matches the beginning of the line</span>
<span class="c"># FYI dollar sign ($) matches the end</span>
<span class="nv">$ </span>grep <span class="s1">&#39;^#&#39;</span> /vol1/opt/data/lamina.bed
</pre></div>
</div>
<p>To find any line that <em>does not</em> start with &#8220;chr&#8221;:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># the -v flag inverts the match (grep &quot;not&quot; [pattern])</span>
<span class="nv">$ </span>grep -v <span class="s1">&#39;^chr&#39;</span> /vol1/opt/data/lamina.bed
</pre></div>
</div>
<p>Beware of using <tt class="docutils literal"><span class="pre">grep</span></tt> to find patterns that might be partial matches:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># this will match chr1, chr10, chr11 etc.</span>
<span class="nv">$ </span>grep chr1 /vol1/opt/data/lamina.bed <span class="p">|</span> cut -f1 <span class="p">|</span> uniq
</pre></div>
</div>
<p>You can find exact matches that are split on words with the <tt class="docutils literal"><span class="pre">-w</span></tt> flag:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># this will only match chr1</span>
<span class="nv">$ </span>grep -w chr1 /vol1/opt/data/lamina.bed <span class="p">|</span> cut -f1 <span class="p">|</span> uniq
</pre></div>
</div>
<p>Beware of using <tt class="docutils literal"><span class="pre">grep</span></tt> to search for numbers:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># finds all strings that match `100`</span>
<span class="nv">$ </span>grep <span class="m">100</span> /vol1/opt/data/lamina.bed <span class="p">|</span> head -n 20

<span class="c"># better, but doesn&#39;t look at numeric value</span>
<span class="nv">$ </span>grep -w <span class="m">100</span> /vol1/opt/data/lamina.bed <span class="p">|</span> head -n 20
</pre></div>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p>If you&#8217;re trying to find numeric values in a file, use <tt class="docutils literal"><span class="pre">awk</span></tt>
instead:</p>
<div class="last highlight-python"><div class="highlight"><pre>$ awk &#39;$2 == 500&#39; /vol1/opt/data/lamina.bed
</pre></div>
</div>
</div>
</div>
<div class="section" id="cluster-access">
<h2>Cluster access<a class="headerlink" href="#cluster-access" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#cluster-access" title="Slides">§</a></h2>
<p>We have set up accounts for the class on our departmental cluster. We will
set up your accounts at the end of class and reset your passwords:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># the -X flag starts an X11 connection</span>
<span class="nv">$ </span>ssh -X username@amc-tesla.ucdenver.pvt

...

<span class="c"># once you are logged in, text your X11 connection with</span>
<span class="nv">$ </span>xeyes
</pre></div>
</div>
</div>
<div class="section" id="cluster-etiquette">
<h2>Cluster etiquette<a class="headerlink" href="#cluster-etiquette" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#cluster-etiquette" title="Slides">§</a></h2>
<p>There are some specific rules you need to know when you&#8217;re operating in a
cluster environment.</p>
<p class="graphviz">
<img src="../_images/graphviz-23b92947c32d4cd35df4fce8fa40a0148601675d.png" alt="digraph cluster {
    &quot;YOU&quot; [shape=box];
    &quot;amc-tesla&quot; [shape=box];
    &quot;filesystem&quot; [shape=box];
    &quot;compute nodes&quot; [shape=box];
    &quot;YOU&quot; -&gt; &quot;amc-tesla&quot;;
    &quot;amc-tesla&quot; -&gt; &quot;filesystem&quot;;
    &quot;amc-tesla&quot; -&gt; &quot;compute nodes&quot;;
}" />
</p>
<div class="admonition important">
<p class="first admonition-title">Important</p>
<p class="last"><strong>DO NOT</strong> run jobs on the head node (amc-tesla). The head node is the
brains of the cluster and it can easily be overextended. Use <tt class="docutils literal"><span class="pre">qlogin</span></tt>
instead.</p>
</div>
</div>
<div class="section" id="example-commands-on-the-cluster">
<h2>Example commands on the cluster<a class="headerlink" href="#example-commands-on-the-cluster" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#example-commands-on-the-cluster" title="Slides">§</a></h2>
<p>Find the size of the file system:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>df -h
</pre></div>
</div>
<p>Find how much space you have allocated:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>quota -h
</pre></div>
</div>
</div>
<div class="section" id="the-queueing-system">
<h2>The queueing system<a class="headerlink" href="#the-queueing-system" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#the-queueing-system" title="Slides">§</a></h2>
<p>First you will grab a single CPU from the queueing system so that you can
work without affecting the head node. We use <tt class="docutils literal"><span class="pre">qlogin</span></tt> for this:</p>
<div class="highlight-bash"><div class="highlight"><pre>jhessel@amc-tesla ~
$ qlogin

Job &lt;492536&gt; is submitted to queue &lt;interactive&gt;.
&lt;&lt;ssh X11 forwarding job&gt;&gt;
&lt;&lt;Waiting for dispatch ...&gt;&gt;
&lt;&lt;Starting on compute00&gt;&gt;

jhessel@compute00 ~
$
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The host in the prompt changed from <tt class="docutils literal"><span class="pre">amc-tesla</span></tt> to <tt class="docutils literal"><span class="pre">compute00</span></tt>.</p>
</div>
<p>You can now execute long-running processes without worry of affecting the
cluster. Type <tt class="docutils literal"><span class="pre">exit</span></tt> to return back to your head node login.</p>
<p>The cluster uses a queueing system that will run jobs that you submit to
it. You can write a small test script to see how the system works. First,
write this into a run.sh file:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">#!/usr/bin/env bash</span>

<span class="c">#BSUB -J sleeper</span>
<span class="c">#BSUB -e %J.err</span>
<span class="c">#BSUB -o %J.out</span>

sleep 20
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">#BSUB</span></tt> lines are comments, but are read by the <tt class="docutils literal"><span class="pre">bsub</span></tt> program to
identify features associated with your job.</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">-J</span></tt> sets the job&#8217;s name</li>
<li><tt class="docutils literal"><span class="pre">%J</span></tt> is a unique job ID that is set when you run the job.</li>
<li><tt class="docutils literal"><span class="pre">-e</span></tt> and <tt class="docutils literal"><span class="pre">-o</span></tt> set the filenames for stderr and stdout from the job</li>
</ul>
<p>Now you can submit the script to the queuing system. As soon as you submit
it, you can check on its progress:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bsub &lt; run.sh
<span class="nv">$ </span>bjobs
</pre></div>
</div>
<p>After the job finishes, you should see two new files that end
<cite>.out</cite> and <cite>.err</cite>; these stdout and stderr from the running job.
Look at the contents of those files so you know what is in
each one.</p>
</div>
<div class="section" id="killing-jobs">
<h2>Killing jobs<a class="headerlink" href="#killing-jobs" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#killing-jobs" title="Slides">§</a></h2>
<p>Sometimes you need to kill your jobs. You can kill specific jobs using
their job ID numbers, obtained from checking <tt class="docutils literal"><span class="pre">bjobs</span></tt>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bkill &lt;jobid&gt;
</pre></div>
</div>
<p>You can also kill <strong>all</strong> of your jobs at once:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bkill 0
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last"><tt class="docutils literal"><span class="pre">bkill</span> <span class="pre">0</span></tt> is dangerous – it will wipe out all of your jobs. If
you have long-running jobs that you forgot about, you will kill them
too if you are not careful!</p>
</div>
</div>
<div class="section" id="other-cluster-specific-commands">
<h2>Other cluster-specific commands<a class="headerlink" href="#other-cluster-specific-commands" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#other-cluster-specific-commands" title="Slides">§</a></h2>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bhosts     <span class="c"># hosts in the cluster</span>
<span class="nv">$ </span>man bhosts <span class="c"># bsub man page</span>
<span class="nv">$ </span>bqueues    <span class="c"># available queues</span>
<span class="nv">$ </span>lsload     <span class="c"># check load values for all hosts</span>
</pre></div>
</div>
</div>
<div class="section" id="questions">
<h2>Questions<a class="headerlink" href="#questions" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#questions" title="Slides">§</a></h2>
<ol class="arabic">
<li><p class="first">Check the <tt class="docutils literal"><span class="pre">.err</span></tt> files from the run. What information do they
contain? What does this tell you about your starting sequences?</p>
</li>
<li><p class="first">Find out how you would modify the <tt class="docutils literal"><span class="pre">bowtie2</span></tt> command to write out the
unaligned reads into a new file. Re-run the analysis to report those
reads.</p>
</li>
<li><p class="first">Modify the <tt class="docutils literal"><span class="pre">awk</span></tt> command in the script to print out a valid BED4
format:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">chrom</span> <span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span> <span class="n">start</span> <span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span> <span class="n">end</span> <span class="o">&lt;</span><span class="n">tab</span><span class="o">&gt;</span> <span class="n">count</span>
</pre></div>
</div>
</li>
<li><p class="first">Find out how many unique UMI sequences are associted with each
chromosomal coordinate (note: not as easy).</p>
</li>
</ol>
</div>
<div class="section" id="more-exercises">
<h2>More exercises<a class="headerlink" href="#more-exercises" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#more-exercises" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>use <tt class="docutils literal"><span class="pre">grep</span></tt> to identify lines in lamina.bed where the second field
(start) begins with <tt class="docutils literal"><span class="pre">100</span></tt>.</li>
<li>use <tt class="docutils literal"><span class="pre">grep</span></tt> to identify lines in lamina.bed where the third field
(end) ends with 99 .</li>
<li>use <tt class="docutils literal"><span class="pre">grep</span></tt> with its <tt class="docutils literal"><span class="pre">-w</span></tt> flag to count the number of &#8216;chr1&#8217;
records in lamina.bed.</li>
<li>use <tt class="docutils literal"><span class="pre">grep</span></tt> to count how many fastq records are in the
/vol1/opt/data/t_R1.fastq.gz file (fastq records begin with an
&#8216;&#64;&#8217; symbol)</li>
<li>login to amc-tesla. use <tt class="docutils literal"><span class="pre">grep</span></tt> to count the number of fastq records
in /vol1/opt/data/SP1.fq.gz</li>
</ol>
</div>
</div>
<div class="section" id="bedtools">
<h1>BEDTools<a class="headerlink" href="#bedtools" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#bedtools" title="Slides">§</a></h1>
<div class="section" id="id1">
<h2>Goals<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#id1" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Introduce the BEDTools suite of tools.</li>
<li>Understand why using BEDTools is needed.</li>
<li>Practice common operations on BED files with BEDTools.</li>
</ol>
</div>
<div class="section" id="bedtools-overview">
<h2>BEDTools Overview<a class="headerlink" href="#bedtools-overview" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#bedtools-overview" title="Slides">§</a></h2>
<p>BEDTools will be one of the tools with the best return on investment. For
example, to extract out <strong>all genes that overlap a CpG island</strong>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -u -a genes.hg19.bed.gz -b cpg.bed.gz <span class="se">\</span>
                                 &gt; genes-in-islands.bed
</pre></div>
</div>
<p><a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/intersect.html#intersect" title="(in bedtools v2.22.1)"><em class="xref std std-ref">intersect</em></a> is a bedtools tool. It follows a
common pattern in bedtools that the query file is specified after the
<tt class="docutils literal"><span class="pre">-a</span></tt> flag and the <em>subject</em> file after the <tt class="docutils literal"><span class="pre">-b</span></tt> flag</p>
</div>
<div class="section" id="bedtools-utility">
<h2>BEDTools Utility<a class="headerlink" href="#bedtools-utility" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#bedtools-utility" title="Slides">§</a></h2>
<p>Finding all overlaps between a pair of BED files naively in python would look like:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">parse_bed</span><span class="p">(</span><span class="s">&#39;a.bed&#39;</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">parse_bed</span><span class="p">(</span><span class="s">&#39;b.bed&#39;</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">overlaps</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">):</span>
            <span class="c"># do stuff</span>
</pre></div>
</div>
<p>If <em>&#8216;a.bed&#8217;</em> has 10K entries and <em>&#8216;b.bed&#8217;</em> has 100K entries, this would
involved checking for overlaps <strong>1 billion times</strong>. That will be slow.</p>
<p>BEDTools uses an indexing scheme that reduces the number of tests
dramatically.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">See the original BEDTools paper for more information:
<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/26/6/841.full">http://bioinformatics.oxfordjournals.org/content/26/6/841.full</a></p>
</div>
<ul class="simple">
<li>Fast: faster than intersect code you will write</li>
<li>Terse: syntax is terse, but readable</li>
<li>Formats: handles BED, VCF and GFF formats (gzip&#8217;ed or not)</li>
<li>Special Cases: handles stranded-ness, 1-base overlaps, abutted intervals,
etc. (likely to be bugs if you do code in manually)</li>
</ul>
</div>
<div class="section" id="bedtools-commands">
<h2>BEDTools Commands<a class="headerlink" href="#bedtools-commands" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#bedtools-commands" title="Slides">§</a></h2>
<p>To see all available BEDTools commands, type</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools
</pre></div>
</div>
<p>The most commonly used BEDtools are:</p>
<ul class="simple">
<li><a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/intersect.html#intersect" title="(in bedtools v2.22.1)"><em class="xref std std-ref">intersect</em></a></li>
<li><a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/genomecov.html#genomecov" title="(in bedtools v2.22.1)"><em class="xref std std-ref">genomecov</em></a></li>
<li><a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/closest.html#closest" title="(in bedtools v2.22.1)"><em class="xref std std-ref">closest</em></a></li>
<li><a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/map.html#map" title="(in bedtools v2.22.1)"><em class="xref std std-ref">map</em></a></li>
</ul>
</div>
<div class="section" id="bedtools-documentation">
<h2>BEDTools Documentation<a class="headerlink" href="#bedtools-documentation" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#bedtools-documentation" title="Slides">§</a></h2>
<p>The BEDTools documentation is quite good and ever improving.</p>
<p>See the documentation for <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/intersect.html#intersect" title="(in bedtools v2.22.1)"><em class="xref std std-ref">intersect</em></a> with:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect
</pre></div>
</div>
<dl class="docutils">
<dt>The online HTML help is also good and includes pictures:</dt>
<dd><a class="reference external" href="https://bedtools.readthedocs.org/en/latest/content/tools/intersect.html">https://bedtools.readthedocs.org/en/latest/content/tools/intersect.html</a></dd>
</dl>
</div>
<div class="section" id="bedtools-intersect">
<h2>BEDTools intersect<a class="headerlink" href="#bedtools-intersect" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#bedtools-intersect" title="Slides">§</a></h2>
<p>Have a browser window open to <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/intersect.html#intersect" title="(in bedtools v2.22.1)"><em class="xref std std-ref">BEDTools intersect documentation</em></a>.
It will likely be the BEDTools function that you use the most. It has a lot of
options.</p>
<img alt="http://bedtools.readthedocs.org/en/latest/_images/intersect-glyph.png" src="http://bedtools.readthedocs.org/en/latest/_images/intersect-glyph.png" />
<p>&#8220;-v&#8221; means (like grep) include all intervals from <cite>-a</cite> that do not overlap
intervals in <cite>-b</cite></p>
</div>
<div class="section" id="example-files">
<h2>Example Files<a class="headerlink" href="#example-files" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#example-files" title="Slides">§</a></h2>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>cat a.bed
chr1    <span class="m">10</span>  <span class="m">20</span>  a1  <span class="m">1</span>   +
chr1    <span class="m">100</span> <span class="m">200</span> a2  <span class="m">2</span>   -

<span class="nv">$ </span>cat b.bed
chr1    <span class="m">20</span>  <span class="m">30</span>  b1  <span class="m">1</span>   +
chr1    <span class="m">90</span>  <span class="m">101</span> b2  <span class="m">2</span>   -
chr1    <span class="m">100</span> <span class="m">110</span> b3  <span class="m">3</span>   +
chr1    <span class="m">200</span> <span class="m">210</span> b4  <span class="m">4</span>   +
</pre></div>
</div>
<p>What will happen if you intersect those files?
For example, the <em>a.bed</em> region <cite>chr1:100-200</cite> overlaps:</p>
<div class="highlight-python"><div class="highlight"><pre>chr1:90-101
chr1:100-110
</pre></div>
</div>
<p>from <em>b.bed</em></p>
</div>
<div class="section" id="intersect">
<h2>intersect<a class="headerlink" href="#intersect" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect" title="Slides">§</a></h2>
<p>intersect with default arguments means <strong>extract chunks of `-a` that overlap
regions in `-b`</strong></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a a.bed -b b.bed
chr1    <span class="m">100</span> <span class="m">101</span> a2  <span class="m">2</span>   -
chr1    <span class="m">100</span> <span class="m">110</span> a2  <span class="m">2</span>   -
</pre></div>
</div>
<p>Here is the original interval from <em>a.bed</em>:</p>
<div class="highlight-python"><div class="highlight"><pre>chr1        100     200     a2      2       -
</pre></div>
</div>
<p>And the overlapping intervals from <em>b.bed</em>:</p>
<div class="highlight-python"><div class="highlight"><pre>chr1        90      101     b2      2       -
chr1        100     110     b3      3       +
</pre></div>
</div>
</div>
<div class="section" id="intersect-wa">
<h2>intersect -wa<a class="headerlink" href="#intersect-wa" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-wa" title="Slides">§</a></h2>
<p>Often, we want the <em>entire interval from -a if it overlaps any interval in -b</em></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a a.bed -b b.bed -wa
chr1    <span class="m">100</span> <span class="m">200</span> a2  <span class="m">2</span>   -
chr1    <span class="m">100</span> <span class="m">200</span> a2  <span class="m">2</span>   -
</pre></div>
</div>
<p>We can get that uniquely with (-u)</p>
</div>
<div class="section" id="intersect-wo">
<h2>intersect -wo<a class="headerlink" href="#intersect-wo" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-wo" title="Slides">§</a></h2>
<p>We can see which intervals in <em>-b</em> are associated with <em>-a</em></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a a.bed -b b.bed -wo
chr1  <span class="m">100</span>  <span class="m">200</span>  a2  <span class="m">2</span>  -  chr1  <span class="m">90</span>  <span class="m">101</span>  b2  <span class="m">2</span>  -  1
chr1  <span class="m">100</span>  <span class="m">200</span>  a2  <span class="m">2</span>  -  chr1  <span class="m">100</span>  <span class="m">110</span>  b3  <span class="m">3</span>  +  10
</pre></div>
</div>
</div>
<div class="section" id="intersect-exercise">
<h2>intersect exercise<a class="headerlink" href="#intersect-exercise" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-exercise" title="Slides">§</a></h2>
<p>What happens if you reverse the arguments? E.g. instead of:</p>
<div class="highlight-python"><div class="highlight"><pre>-a a.bed -b b.bed
</pre></div>
</div>
<p>use:</p>
<div class="highlight-python"><div class="highlight"><pre>-b a.bed -a b.bed
</pre></div>
</div>
<p>Try that with no extra flags, with -u, -wa, -wo.</p>
<p>How does it compare to the original?</p>
</div>
<div class="section" id="intersect-c">
<h2>intersect -c<a class="headerlink" href="#intersect-c" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-c" title="Slides">§</a></h2>
<p>We can count overlaps for each interval in <em>-a</em> with those in <em>-b</em> with</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a a.bed -b b.bed -c
chr1        <span class="m">10</span>      <span class="m">20</span>      a1      <span class="m">1</span>       +       0
chr1        <span class="m">100</span>     <span class="m">200</span>     a2      <span class="m">2</span>       -       2
</pre></div>
</div>
<p>This is our original <cite>a.bed</cite> with an <strong>additional column indicating number of
overlaps</strong> with <cite>b.bed</cite></p>
</div>
<div class="section" id="intersect-v">
<h2>intersect -v<a class="headerlink" href="#intersect-v" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-v" title="Slides">§</a></h2>
<p>Extract intervals in <cite>a.bed</cite> that do not overlap any interval in <cite>b.bed</cite></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a a.bed -b b.bed -v
chr1        <span class="m">10</span>      <span class="m">20</span>      a1      <span class="m">1</span>       +
</pre></div>
</div>
<p>Extract intervals in <cite>b.bed</cite> that do not overlap any interval in <cite>a.bed</cite></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a b.bed -b a.bed -v
chr1        <span class="m">20</span>      <span class="m">30</span>      b1      <span class="m">1</span>       +
chr1        <span class="m">200</span>     <span class="m">210</span>     b4      <span class="m">4</span>       +
</pre></div>
</div>
</div>
<div class="section" id="intersect-summary">
<h2>Intersect Summary<a class="headerlink" href="#intersect-summary" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-summary" title="Slides">§</a></h2>
<ul class="simple">
<li>fragments of <cite>a</cite> that overlap <cite>b</cite>:
<cite>intersect -a a.bed -b b.bed</cite></li>
<li>complete regions of <cite>a</cite> that overlap <cite>b</cite>:
<cite>intersect -a a.bed -b b.bed -u</cite></li>
<li>intervals of <cite>b</cite> as well as <cite>a</cite>:
<cite>intersect -a a.bed -b b.bed -wo</cite></li>
<li>number of times each <cite>a</cite> overlaps <cite>b</cite>:
<cite>intersect -a a.bed -b b.bed -c</cite></li>
<li>intervals of <cite>a</cite> that do not overlap <cite>b</cite>:
<cite>intersect -a a.bed -b b.bed -v</cite></li>
</ul>
</div>
<div class="section" id="exercises-or-other-tools">
<h2>Exercises (Or Other Tools)<a class="headerlink" href="#exercises-or-other-tools" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#exercises-or-other-tools" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>zless <a class="reference download internal" href="../_downloads/cpg.bed.gz"><tt class="xref download docutils literal"><span class="pre">cpg.bed.gz</span></tt></a> and <a class="reference download internal" href="../_downloads/genes.hg19.bed.gz"><tt class="xref download docutils literal"><span class="pre">genes.hg19.bed.gz</span></tt></a></li>
<li>Extract the CpG islands that touch any gene [<strong>24611</strong>]</li>
<li>Extract CpG islands that do not touch any gene [<strong>7012</strong>]</li>
<li>Extract (uniquely) all of each CpG Island that touches any gene [<strong>21679</strong>]</li>
<li>Extract CpG&#8217;s that are completely contained within a gene (look at the help
for a flag to indicate that you want the fraction of overlap to be 1 (for 100 %). [<strong>10714</strong>]</li>
<li>Report genes that overlap any CpG island. [<strong>16908</strong>]</li>
<li>Report genes that overlap more than 1 CpG Island (use -c and awk). [<strong>3703</strong>].</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">as you are figuring these out, make sure to pipe the output to less or head</p>
</div>
</div>
<div class="section" id="other-reading">
<h2>Other Reading<a class="headerlink" href="#other-reading" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#other-reading" title="Slides">§</a></h2>
<ul class="simple">
<li>Check out the online <a class="reference external" href="https://bedtools.readthedocs.org/en/latest/content/tools/intersect.html">documentation</a>.</li>
<li>A <a class="reference external" href="http://quinlanlab.org/tutorials/cshl2013/bedtools.html">tutorial</a> by the author of BEDTools</li>
</ul>
</div>
<div class="section" id="intersect-bam">
<h2>Intersect Bam<a class="headerlink" href="#intersect-bam" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-bam" title="Slides">§</a></h2>
<p>We have seen that <cite>intersect &lt;bedtools:intersect&gt;</cite> takes <cite>-a</cite> and <cite>-b</cite>
arguments. It can also intersect against an alignment BAM file by using <cite>-abam</cite>
in place of <cite>-a</cite></p>
<p>e.g:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect <span class="se">\</span>
    -abam experiment.bam <span class="se">\</span>
    -b target-regions.bed <span class="se">\</span>
    &gt; on-target.bam
</pre></div>
</div>
</div>
<div class="section" id="intersect-strand">
<h2>Intersect Strand<a class="headerlink" href="#intersect-strand" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#intersect-strand" title="Slides">§</a></h2>
<p>From the <a class="reference external" href="https://bedtools.readthedocs.org/en/latest/content/tools/intersect.html">help</a> ,
one can see that intersect can consider strand. For example if both files have a
strand field then</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools intersect -a a.bed -b b.bed -s
</pre></div>
</div>
<p>Will only consider as overlapping those intervals in <cite>a.bed</cite> that have the same
strand as <cite>b.bed</cite>.</p>
</div>
<div class="section" id="closest">
<h2>Closest<a class="headerlink" href="#closest" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#closest" title="Slides">§</a></h2>
<p>with <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/intersect.html#intersect" title="(in bedtools v2.22.1)"><em class="xref std std-ref">intersect</em></a> we can only get overlapping
intervals. <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/closest.html#closest" title="(in bedtools v2.22.1)"><em class="xref std std-ref">closest</em></a> reports the nearest interval even
if it&#8217;s not overlapping.</p>
<p>Example: report the nearest CpG to each gene as long as it is within 5KB.</p>
<div class="highlight-bash"><div class="highlight"><pre>bedtools closest <span class="se">\</span>
    -a genes.hg19.bed.gz <span class="se">\</span>
    -b cpg.bed.gz -d <span class="se">\</span>
    <span class="p">|</span> awk <span class="s1">&#39;$NF &lt;= 5000&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="map">
<h2>Map<a class="headerlink" href="#map" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#map" title="Slides">§</a></h2>
<p>For each CpG print the sum of the values (4th column) of overlapping intervals from
lamina.bed (and filter out those with no overlap using awk)</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools map <span class="se">\</span>
    -a cpg.bed.gz <span class="se">\</span>
    -b /vol1/opt/data/lamina.bed <span class="se">\</span>
    -c <span class="m">4</span> -o sum <span class="se">\</span>
    <span class="p">|</span> awk <span class="s1">&#39;$5 != &quot;.&quot;&#39;</span>
</pre></div>
</div>
<p>Other <em>-o</em> perations include <strong>min</strong>, <strong>max</strong>, <strong>mean</strong>, <strong>median</strong>, <strong>concat</strong></p>
</div>
<div class="section" id="sorted">
<h2>Sorted<a class="headerlink" href="#sorted" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#sorted" title="Slides">§</a></h2>
<p>When you start dealing with larger data-files. Look at the <cite>-sorted</cite> flag.
For example in <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/intersect.html#intersect" title="(in bedtools v2.22.1)"><em class="xref std std-ref">intersect</em></a>.</p>
<ul class="simple">
<li>Uses less memory</li>
<li>Faster</li>
</ul>
<p>Takes advantage of sorted chromosome, positions in both files so it doesn&#8217;t have
to create an index.</p>
<img alt="http://bedtools.readthedocs.org/en/latest/_images/speed-comparo.png" src="http://bedtools.readthedocs.org/en/latest/_images/speed-comparo.png" />
</div>
<div class="section" id="genomecov">
<h2>Genomecov<a class="headerlink" href="#genomecov" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#genomecov" title="Slides">§</a></h2>
<p>Get coverage of intervals in BED by BAM</p>
<img alt="https://bedtools.readthedocs.org/en/latest/_images/genomecov-glyph.png" src="https://bedtools.readthedocs.org/en/latest/_images/genomecov-glyph.png" />
<p>Usually want the last option <cite>-bg -split</cite></p>
</div>
<div class="section" id="encode">
<h2>ENCODE<a class="headerlink" href="#encode" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#encode" title="Slides">§</a></h2>
<p>The Human Genome Project was finished, giving us a list of human genes and their
locations. Unfortunately, we still had no idea how they were regulated. If only
there was an <a class="reference external" href="http://www.sciencemag.org.hsl-ezproxy.ucdenver.edu/content/306/5696/636.full">ENCyclopedia Of Dna Elements</a>…</p>
<p>Advantages: massive amounts of information on key cell lines, reproducible
experiments, public data access, technology development.</p>
</div>
<div class="section" id="encode-project-cell-lines">
<h2>ENCODE Project Cell Lines<a class="headerlink" href="#encode-project-cell-lines" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#encode-project-cell-lines" title="Slides">§</a></h2>
<p>Tier 1: GM12878 (EBV-transformed lymphoblast), K562 (CML lymphoblast), H1-hESC</p>
<p>Tier 2: HeLa-S3 (cervical cancer), HepG2 (liver carcinoma), HUVEC (umbilical vein)</p>
<p>Tier 2.5: SKNSH (neuroblastoma), IMR90 (lung fibroblast), A549 (lung carcinoma),
MCF7 (breast carcinoma), LHCN (myoblast), CD14+, CD20+</p>
<p><a class="reference external" href="http://genome.ucsc.edu/ENCODE/cellTypes.html">link</a> (this page also has very useful
links to cell culture protocols)</p>
</div>
<div class="section" id="experiments">
<h2>Experiments<a class="headerlink" href="#experiments" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#experiments" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>ChIP-seq: Histone marks, transcription factors</li>
<li>Chromatin structure: DNaseI-seq, FAIRE, 5C/Hi-C</li>
<li>RNA expression: mRNA-seq, GENCODE gene predictions</li>
<li>Data Integration: Segway / ChromHMM integration of functional data</li>
</ol>
</div>
<div class="section" id="common-file-formats">
<h2>Common File Formats<a class="headerlink" href="#common-file-formats" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#common-file-formats" title="Slides">§</a></h2>
<ul class="simple">
<li>FASTQ: Raw sequencing data. <cite>[link] &lt;http://maq.sourceforge.net/fastq.shtml&gt;</cite></li>
<li>SAM/BAM: Aligned sequence data <cite>[link] &lt;http://samtools.github.io/hts-specs/SAMv1.pdf&gt;</cite></li>
<li>Bed/bigBed: List of genomic regions <cite>[link] &lt;http://genome.ucsc.edu/FAQ/FAQformat.html#format1&gt;</cite></li>
<li>Bedgraph/Wig/bigWig: Continuous signal <cite>[link] &lt;http://genome.ucsc.edu/goldenPath/help/bedgraph.html&gt;</cite></li>
</ul>
<p>Many other formats are described on this <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html">page</a></p>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#references" title="Slides">§</a></h2>
<p>Completion of the entire project, and a ton of papers:
<a class="reference external" href="http://www.nature.com/nature/journal/v489/n7414/index.html">Nature</a>,
<a class="reference external" href="http://genome.cshlp.org/content/22/9.toc">Genome Research</a>,
<a class="reference external" href="http://genomebiology.com/content/13/9">Genome Biology</a>,</p>
</div>
<div class="section" id="how-to-access-encode-data">
<h2>How to Access ENCODE Data<a class="headerlink" href="#how-to-access-encode-data" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-5.html#how-to-access-encode-data" title="Slides">§</a></h2>
<p>The <a class="reference external" href="https://www.encodeproject.org/">ENCODE project page</a> is the portal
to all of the ENCODE data.</p>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="../_sources/Classes/class-5.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2014, Jay Hesselberth.<br/>
      Last updated on Feb 13, 2015.<br/>
    </p>
  </div>
</footer>
  </body>
</html>