<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Class 6 : Genomic Data Vignette: ChIP-seq &mdash; Genome Analysis Workshop 0.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/flatly/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Class 7 : BEDTools vignette" href="class-07.html" />
    <link rel="prev" title="Class 5 : BEDTools" href="class-05.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Genome Analysis Workshop</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="http://hesselberthlab.github.io">Hesselberth lab</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Genome Analysis Workshop <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Syllabus.html">Syllabus</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../Classes.html">Class list</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Exercises.html">Exercises</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Problem_Sets.html">Problem Sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Miscellaneous.html">Miscellaneous</a></li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul>
<li><a class="reference internal" href="#">Class 6 : Genomic Data Vignette: ChIP-seq</a><ul>
<li><a class="reference internal" href="#goals">Goals</a></li>
<li><a class="reference internal" href="#encode">ENCODE</a></li>
<li><a class="reference internal" href="#encode-project-cell-lines">ENCODE Project Cell Lines</a></li>
<li><a class="reference internal" href="#experiments">Experiments</a></li>
<li><a class="reference internal" href="#common-file-formats">Common File Formats</a></li>
<li><a class="reference internal" href="#references">References</a></li>
<li><a class="reference internal" href="#how-to-access-encode-data">How to Access ENCODE Data</a></li>
<li><a class="reference internal" href="#chromatin-immunoprecipitation-overview">Chromatin Immunoprecipitation Overview</a></li>
<li><a class="reference internal" href="#chip-seq-analysis-workflow">ChIP-seq analysis workflow</a></li>
<li><a class="reference internal" href="#chip-seq-data">ChIP-seq data</a></li>
<li><a class="reference internal" href="#peak-calling">Peak calling</a></li>
<li><a class="reference internal" href="#identify-sequence-motifs-in-enriched-regions">Identify sequence motifs in enriched regions</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    <div class="col-md-9">
      
  <div class="section" id="class-6-genomic-data-vignette-chip-seq">
<h1>Class 6 : Genomic Data Vignette: ChIP-seq<a class="headerlink" href="#class-6-genomic-data-vignette-chip-seq" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#class-6-genomic-data-vignette-chip-seq" title="Slides">§</a></h1>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">2015 Feb 12 Thurs</td>
</tr>
<tr class="field-even field"><th class="field-name">Last updated:</th><td class="field-body">March 06, 2015</td>
</tr>
</tbody>
</table>
<div class="section" id="goals">
<h2>Goals<a class="headerlink" href="#goals" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#goals" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>Learn workflows for analyzing ChIP-seq data</li>
<li>Derive DNA sequence motifs from ChIP-seq peaks</li>
</ol>
</div>
<div class="section" id="encode">
<h2>ENCODE<a class="headerlink" href="#encode" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#encode" title="Slides">§</a></h2>
<p>The Human Genome Project was finished, giving us a list of human genes and their
locations. Unfortunately, we still had no idea how they were regulated. If only
there was an <a class="reference external" href="http://www.sciencemag.org.hsl-ezproxy.ucdenver.edu/content/306/5696/636.full">ENCyclopedia Of Dna Elements</a>…</p>
<p>Advantages: massive amounts of information on key cell lines, reproducible
experiments, public data access, technology development.</p>
</div>
<div class="section" id="encode-project-cell-lines">
<h2>ENCODE Project Cell Lines<a class="headerlink" href="#encode-project-cell-lines" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#encode-project-cell-lines" title="Slides">§</a></h2>
<p>Tier 1: GM12878 (EBV-transformed lymphoblast), K562 (CML lymphoblast), H1-hESC</p>
<p>Tier 2: HeLa-S3 (cervical cancer), HepG2 (liver carcinoma), HUVEC (umbilical vein)</p>
<p>Tier 2.5: SKNSH (neuroblastoma), IMR90 (lung fibroblast), A549 (lung carcinoma),
MCF7 (breast carcinoma), LHCN (myoblast), CD14+, CD20+</p>
<p><a class="reference external" href="http://genome.ucsc.edu/ENCODE/cellTypes.html">link</a> (this page also has very useful
links to cell culture protocols)</p>
</div>
<div class="section" id="experiments">
<h2>Experiments<a class="headerlink" href="#experiments" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#experiments" title="Slides">§</a></h2>
<ol class="arabic simple">
<li>ChIP-seq: Histone marks, transcription factors</li>
<li>Chromatin structure: DNaseI-seq, FAIRE, 5C/Hi-C</li>
<li>RNA expression: mRNA-seq, GENCODE gene predictions</li>
<li>Data Integration: Segway / ChromHMM integration of functional data</li>
</ol>
</div>
<div class="section" id="common-file-formats">
<h2>Common File Formats<a class="headerlink" href="#common-file-formats" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#common-file-formats" title="Slides">§</a></h2>
<ul class="simple">
<li>FASTQ: Raw sequencing data. <cite>[link] &lt;http://maq.sourceforge.net/fastq.shtml&gt;</cite></li>
<li>SAM/BAM: Aligned sequence data <cite>[link] &lt;http://samtools.github.io/hts-specs/SAMv1.pdf&gt;</cite></li>
<li>Bed/bigBed: List of genomic regions <cite>[link] &lt;http://genome.ucsc.edu/FAQ/FAQformat.html#format1&gt;</cite></li>
<li>Bedgraph/Wig/bigWig: Continuous signal <cite>[link] &lt;http://genome.ucsc.edu/goldenPath/help/bedgraph.html&gt;</cite></li>
</ul>
<p>Many other formats are described on this <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html">page</a></p>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#references" title="Slides">§</a></h2>
<p>Completion of the entire project, and a ton of papers:
<a class="reference external" href="http://www.nature.com/nature/journal/v489/n7414/index.html">Nature</a>,
<a class="reference external" href="http://genome.cshlp.org/content/22/9.toc">Genome Research</a>,
<a class="reference external" href="http://genomebiology.com/content/13/9">Genome Biology</a>,</p>
</div>
<div class="section" id="how-to-access-encode-data">
<h2>How to Access ENCODE Data<a class="headerlink" href="#how-to-access-encode-data" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#how-to-access-encode-data" title="Slides">§</a></h2>
<p>The <a class="reference external" href="https://www.encodeproject.org/">ENCODE project page</a> is the portal
to all of the ENCODE data.</p>
</div>
<div class="section" id="chromatin-immunoprecipitation-overview">
<h2>Chromatin Immunoprecipitation Overview<a class="headerlink" href="#chromatin-immunoprecipitation-overview" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#chromatin-immunoprecipitation-overview" title="Slides">§</a></h2>
<p>Chromatin Immunoprecipitation is used to determine where a protein of
interest binds on a chromatin template <a class="reference internal" href="#park-chipseq" id="id1">[Park_Chipseq]</a>.</p>
<table class="docutils citation" frame="void" id="park-chipseq" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[Park_Chipseq]</a></td><td><a class="reference external" href="http://www.nature.com/nrg/journal/v10/n10/full/nrg2641.html">http://www.nature.com/nrg/journal/v10/n10/full/nrg2641.html</a></td></tr>
</tbody>
</table>
<img alt="../_images/chip-workflow.png" src="../_images/chip-workflow.png" />
<img alt="../_images/chip-data.png" src="../_images/chip-data.png" />
</div>
<div class="section" id="chip-seq-analysis-workflow">
<h2>ChIP-seq analysis workflow<a class="headerlink" href="#chip-seq-analysis-workflow" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#chip-seq-analysis-workflow" title="Slides">§</a></h2>
<p>A general workflow for visualizing ChIP-seq data (and many other types of
data) is:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Operation</th>
<th class="head">File formats</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Align reads to reference genome</td>
<td><tt class="docutils literal"><span class="pre">FASTQ</span> <span class="pre">~~&gt;</span> <span class="pre">BAM</span></tt></td>
</tr>
<tr class="row-odd"><td>Generate coverage plots</td>
<td><tt class="docutils literal"><span class="pre">BAM</span> <span class="pre">~~&gt;</span> <span class="pre">bedGraph</span></tt></td>
</tr>
<tr class="row-even"><td>Call peaks</td>
<td><tt class="docutils literal"><span class="pre">BAM</span> <span class="pre">~~&gt;</span> <span class="pre">BED</span></tt></td>
</tr>
<tr class="row-odd"><td>Make binary files for UCSC display</td>
<td><tt class="docutils literal"><span class="pre">bedGraph</span> <span class="pre">~~&gt;</span> <span class="pre">bigWig</span></tt>, <tt class="docutils literal"><span class="pre">BED</span> <span class="pre">~~&gt;</span> <span class="pre">bigBed</span></tt></td>
</tr>
<tr class="row-even"><td>Identify motifs</td>
<td><tt class="docutils literal"><span class="pre">BED</span> <span class="pre">~~&gt;</span> <span class="pre">FASTA</span> <span class="pre">~~&gt;</span> <span class="pre">TXT</span> <span class="pre">/</span> <span class="pre">HTML</span></tt></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="chip-seq-data">
<h2>ChIP-seq data<a class="headerlink" href="#chip-seq-data" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#chip-seq-data" title="Slides">§</a></h2>
<p>Look at some human ChIP-seq data <a class="footnote-reference" href="#id3" id="id2">[1]</a>.</p>
<table class="docutils footnote" frame="void" id="id3" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[1]</a></td><td>Genome Browser Session
<a class="reference external" href="http://goo.gl/WfJxcM">http://goo.gl/WfJxcM</a></td></tr>
</tbody>
</table>
<p>(We&#8217;ll talk more in depth about ChIP-Seq workflows in the future,
but for now, just a brief introdcution to a few commands you can use to
work on ChIP-Seq data (and pset3)).</p>
</div>
<div class="section" id="peak-calling">
<h2>Peak calling<a class="headerlink" href="#peak-calling" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#peak-calling" title="Slides">§</a></h2>
<p>There are several available software packages for identifying regions
enriched in your IP experiment (i.e. peaks). We will use macs2 here.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># minimal macs2 command</span>
<span class="nv">$ </span>macs2 callpeak --treatment &lt;aln.bam&gt; --name &lt;exp.name&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="identify-sequence-motifs-in-enriched-regions">
<h2>Identify sequence motifs in enriched regions<a class="headerlink" href="#identify-sequence-motifs-in-enriched-regions" title="Permalink to this headline">¶</a><a class="headerlink" href="../slides/Classes/class-06.html#identify-sequence-motifs-in-enriched-regions" title="Slides">§</a></h2>
<p>You can use meme <a class="footnote-reference" href="#id5" id="id4">[2]</a> to identify over-represented motifs in groups of
sequences (e.g. sequences covered by ChIP peaks).</p>
<p>Use the <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/getfasta.html#getfasta" title="(in bedtools v2.22.1)"><em class="xref std std-ref">bedtools getfasta</em></a> command to fetch
fasta sequences.</p>
<p>Note: meme looks at both strands of a DNA sequence by default.</p>
<table class="docutils footnote" frame="void" id="id5" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id4">[2]</a></td><td>MEME
<a class="reference external" href="http://meme.nbcr.net/meme/">http://meme.nbcr.net/meme/</a></td></tr>
</tbody>
</table>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools getfasta -fi &lt;ref.fa&gt; -bed &lt;peaks.bed&gt; -fo peaks.fa
<span class="nv">$ </span>meme -nmotifs <span class="m">5</span> -minw <span class="m">6</span> -maxw <span class="m">20</span> -dna &lt;peaks.fa&gt;
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>

  </body>
</html>