********
Syllabus
********

:Title: Genome Informatics Workshop
:Course Number: MOLB 7620
:Semester: Spring 2014
:Homepage: http://ucd-bioworkshop.github.io
:Author: Jay R. Hesselberth
:Author: Brent Pedersen
:Organization: University of Colorado School of Medicine
:Address: Department of Biochemistry and Molecular Genetics
:Address: Department of Medicine
:Copyright: 2013-2014 Jay R. Hesselberth, Brent Pedersen
:Copyright: 2013-2014 Regents of the University of Colorado.
:Copyright: All Rights Reserved.
:Last updated: |today|

.. important::

    The course content can be downloaded in PDF form: 
    :download:`PDF Download <_build/pdf/Genome-Informatics-Workshop.pdf>`

Instructor Information
======================

+-------------------------------------+
| **Jay Hesselberth**                 |
+-------------------------------------+
| jay.hesselberth@gmail.com           |
+-------------------------------------+
| **Brent Pedersen**                  |
+-------------------------------------+
| bpederse@gmail.com                  |
+-------------------------------------+
| **Eric Nguyen** (TA)                |
+-------------------------------------+
| eric.2.nguyen@ucdenver.edu          |
+-------------------------------------+
| **Sally Peach** (TA)                |
+-------------------------------------+
| sallypeach@gmail.com                |
+-------------------------------------+

Schedule
========
Mon, Wed, Fri, 2:30 - 4:00 PM

Location
========
Health Sciences Library, Computer Teaching Lab 2

Course Description
==================

The Genome Informatics Workshop is a hands-on tutorial of skills needed to
process large genomics data sets and visualize their results.  The class
is taught from the standpoint of an experimentalist with practical goals
(e.g. to interpret the results of a sequencing-based experiment and gain
biologically meaningful insight).

In the first block, we focus on working in the Linux environment, with
considerable emphasis on Linux command-line tools, Python programming and
the R statistical computing environment. We use publicly available
next-generation DNA sequencing data from the ENCODE project to illustrate
standard approaches for manipulating sequencing data, aligning sequences
to a reference genome, generating coverage plots and displaying them in
the UCSC Genome Browser.

In the second block, we focus on specific analyses used in ENCODE project,
including ChIP-seq, DNase I footprinting, mRNA-seq and genome sequencing
to identify single nucleotide and structural variants.

Course Credits
--------------
This is a 2 credit course.

Texts and Reading Materials
---------------------------

#. **Required**: Learn Python the Hard Way,
    http://learnpythonthehardway.org/book/
#. **Required**: Command Line Crash Course
    http://cli.learncodethehardway.org/book/
#. **Required**: ggplot2: Elegant Graphics for Data Analysis
    http://ggplot2.org/book/
#. **Required**: A Quick Guide to Organizing Computational Biology Projects
    http://www.ploscompbiol.org/article/metrics/info%3Adoi%2F10.1371%2Fjournal.pcbi.1000424

Course Objectives
-----------------
  1. First block (6 weeks, 27 Jan - 28 Feb)

    - Learn to manipulate large sequencing data sets with Linux tools
      and Python programming.

    - Learn to manipulate and visualize data in useful ways with the
      R statistical computing environment.

  2. Second block (6 weeks, 3 Mar - 14 Apr)

    - Learn workflows for experiments in ENCODE: ChIP-seq DNase I
      footprinting, mRNA-seq, variant detection.

    - Learn to visualize data from the ENCODE project in the UCSC Genome
      Browser

Canvas 
======

The course has a Canvas page [#]_ where announcements are made and
problem sets are uploaded

.. [#] https://ucdenver.instructure.com/courses/11079/assignments/syllabus

.. note::

    Be sure to **login** to the Canvas site to be able to see Announcements,
    upload Problem Sets, etc.

Assessment
==========
Progress of individual students will be assessed during the daily exercise
session, weekly problem sets, as well as a final project.

Grading Criteria
----------------
 - 50% participation
 - 35% problem sets (7 sets, 5% each)
 - 15% final project

Specific Dates / Material to be Covered
=======================================

Block 1 
-------

    - Class 1 (M Jan 27) – Introduction to VM, Linux and the shell
    - Class 2 (W Jan 29) – Linux / Utilities
    - Class 3 (F Jan 31) – Linux / Utilities
    - Class 4 (M Feb 3) – Linux / Utilities
    - Class 5 (W Feb 5) – Cluster Usage / Review
    - Class 6 (F Feb 7) – Cluster Usage / Review
    - Class 7 (M Feb 10) – Python
    - Class 8 (W Feb 12) – Python
    - Class 9 (F Feb 14) – Python – Advanced / Review

    **No Class Feb 17 (President's Day)**

    - Class 10 (W Feb 19) – Python – Advanced
    - Class 11 (F Feb 21) – Python – Writing Scripts / Review
    - Class 12 (M Feb 24) – R – Data 
    - Class 13 (W Feb 26) – R – Data
    - Class 14 (F Feb 28) – R – Plotting  / Review
    - Class 15 (M Mar 3) – R – Plotting
    - Class 16 (W Mar 5) – UCSC Genome Browser

Block 2
-------
    - Class 1 (F Mar 7) – UCSC Genome Browser

    **No Class Mar 10-14 (Spring Break)**

    - Class 2 (M Mar 17) – ENCODE – Overview
    - Class 3 (W Mar 19) – ChIP-seq / DNaseI-seq (Read densities)
    - Class 4 (F Mar 21) – ChIP-seq / DNaseI-seq (Peak calling)
    - Class 5 (M Mar 24) – ChIP-seq / DNaseI-seq (Motif finding with MEME)
    - Class 6 (W Mar 26) – BEDtools (Quinlan)
    - Class 7 (F Mar 28) – mRNA-seq (RPKM values and ranking)
    - Class 8 (M Mar 31) – mRNA-seq (Differential expression)
    - Class 9 (W Apr 2) – mRNA-seq (Transcript isoforms)
    - Class 10 (F Apr 4) – Variant detection (Alignment and reprocessing)
    - Class 11 (M Apr 7) – Variant detection (SNP calling; heterozygotes)
    - Class 12 (W Apr 9) – Variant detection (CNV detection)
    - Class 13 (F Apr 11) – Variant detection (VCF files and manipulation)
    - Class 14 (M Apr 14) – Integration / Segway (Hoffman)

.. raw:: pdf

    PageBreak
