********************
Syllabus (MOLB 7620)
********************

General Information
===================

:Title: Genome Informatics Workshop
:Course Number: MOLB 7620
:Semester: Spring 2014
:Homepage: http://ucd-bioworkshop.github.io
:Author: Jay R. Hesselberth
:Author: Brent Pedersen
:Organization: University of Colorado School of Medicine
:Address: Department of Biochemistry and Molecular Genetics
:Address: Department of Medicine
:Copyright: 2013-2014 Jay R. Hesselberth, Brent Pedersen
:Copyright: All Rights Reserved.
:Last updated: |today|

PDF Content
===========

The course content is available as a combined PDF: 
:download:`PDF Download <_build/pdf/Genome-Informatics-Workshop.pdf>`

Instructor Information
======================

.. list-table::
    :widths: 40 40
    :header-rows: 1

    * - Instructor
      - Email
    * - Jay Hesselberth             
      - jay.hesselberth@gmail.com
    * - Brent Pedersen
      - bpederse@gmail.com
    * - Eric Nguyen (TA)
      - eric.2.nguyen@ucdenver.edu
    * - Sally Peach (TA)
      - sallypeach@gmail.com

Schedule
========
Mon, Wed, Fri, 2:30 - 4:00 PM.
:ref:`See specific dates. <syllabus-specific-dates>`

Location
========
Health Sciences Library, Computer Teaching Lab 2

Course Description
==================

**The Genome Informatics Workshop is a hands-on tutorial of skills needed to
process large genomics data sets and visualize their results. The class
is taught from the standpoint of biologist with practical goals
(e.g. to interpret the results of a sequencing-based experiment and gain
biologically meaningful insight).**

We focus on working in the Linux environment, with emphasis on Linux
command-line tools, Python programming and the R statistical computing
environment. We use publicly available next-generation DNA sequencing data
from the ENCODE project to illustrate standard approaches for manipulating
sequencing data, aligning sequences to a reference genome, generating
coverage plots and displaying them in the UCSC Genome Browser. We will
cover specific analyses used in ENCODE project including ChIP-seq, DNase I
footprinting, mRNA-seq and genome sequencing to identify single nucleotide
and structural variants.

Course Credits
--------------
This is a 2 credit course.

Texts and Reading Materials
---------------------------

#. **Required**: Learn Python the Hard Way,
    http://learnpythonthehardway.org/book/
#. **Required**: Command Line Crash Course
    http://cli.learncodethehardway.org/book/
#. **Required**: ggplot2: Elegant Graphics for Data Analysis
    http://ggplot2.org/book/
#. **Required**: A Quick Guide to Organizing Computational Biology Projects
    http://www.ploscompbiol.org/article/metrics/info%3Adoi%2F10.1371%2Fjournal.pcbi.1000424

Course Objectives
-----------------

  - Learn to manipulate large sequencing data sets with Linux tools
    and Python programming.

  - Learn to manipulate and visualize data in useful ways with the
    R statistical computing environment.

  - Learn workflows for experiments in ENCODE: ChIP-seq DNase I
    footprinting, mRNA-seq, variant detection.

  - Learn to visualize data from the ENCODE project in the UCSC Genome
    Browser

Canvas 
======

The course has a Canvas page [#]_ where announcements are made and
problem sets are uploaded

.. [#] https://ucdenver.instructure.com/courses/11079/assignments/syllabus

.. note::

    Be sure to **login** to the Canvas site to be able to see Announcements,
    upload Problem Sets, etc.

Assessment
==========
Progress of individual students will be assessed during the daily exercise
session, weekly problem sets, as well as a final project.

Grading Criteria
----------------
 - 50% participation
 - 35% problem sets (7 sets, 5% each)
 - 15% final project

.. _syllabus-specific-dates:

Specific Dates / Material to be Covered
=======================================

.. list-table::
    :widths: 20 40 80
    :header-rows: 1

    * - Class number
      - Date
      - Topic
    * - Class 1
      - M Jan 27
      - Introduction to VM, Linux and the shell
    * - Class 2 
      - W Jan 29 
      - Linux / Utilities
    * - Class 3 
      - F Jan 31
      - Linux / Utilities
    * - Class 4 
      - M Feb 3
      - Linux / Utilities
    * - Class 5 
      - W Feb 5
      - Cluster Usage / Review
    * - Class 6 
      - F Feb 7
      - Cluster Usage / Review
    * - Class 7 
      - M Feb 10
      - Python
    * - Class 8 
      - W Feb 12
      - Python
    * - Class 9 
      - F Feb 14
      - Python 
    * -
      - **No Class Feb 17 (President's Day)**
      -
    * - Class 10 
      - W Feb 19
      - Python 
    * - Class 11 
      - F Feb 21
      - Python 
    * - Class 12
      - M Feb 24
      - ENCODE Overview
    * - Class 13 
      - W Feb 26
      - BEDtools  
    * - Class 14 
      - F Feb 28
      - ChIP-seq (coverage / peaks / motifs)
    * - Class 15 
      - M Mar 3
      - ChIP-seq / DNaseI-seq (UCSC)
    * - Class 16
      - W Mar 5
      - Genomic analysis vignettes 
    * - Class 17 
      - F Mar 7
      - Genomic analysis vignettes 
    * -
      - **No Class Mar 10-14 (Spring Break)**
      -
    * - Class 18
      - M Mar 17 
      - R data & plotting 
    * - Class 19
      - W Mar 19 
      - R data & plotting 
    * - Class 20
      - F Mar 21 
      - R data & plotting 
    * - Class 21
      - M Mar 24 
      - R data & plotting 
    * - Class 22
      - W Mar 26 
      - R data & plotting 
    * - Class 23
      - F Mar 28 
      - mRNA-seq (FPKM / diff exp)
    * - Class 24
      - M Mar 31 
      - mRNA-seq 
    * - Class 25 
      - W Apr 2 
      - Variant detection - VCF / SNP calling 
    * - Class 26 
      - F Apr 4 
      - Variant detection - Phasing / CNV
    * - Class 27 
      - M Apr 7 
      - TBD
    * - Class 28 
      - W Apr 9 
      - TBD
    * - Class 29 
      - F Apr 11 
      - TBD
    * - Class 30 
      - M Apr 14 
      - Final project presentations

.. raw:: pdf

    PageBreak
