<!--
Google IO 2012/2013 HTML5 Slide Template

Authors: Eric Bidelman <ebidel@gmail.com>
         Luke Mahé <lukem@google.com>

URL: https://code.google.com/p/io-2012-slides
--><!DOCTYPE html>


<html>
<head>
  <title>Problem Set 3 : KEY &mdash; Genome Analysis Workshop 0.2 documentation</title>
  <meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">-->
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0">-->
  <!--This one seems to work all the time, but really small on ipad-->
  <!--<meta name="viewport" content="initial-scale=0.4">-->
  <meta name="apple-mobile-web-app-capable" content="yes">

  <meta name="hieroglyph-title" data-config-title>
  <meta name="hieroglyph-subtitle" data-config-subtitle>
  <meta name="hieroglyph-presenter" data-config-presenter>

  
  <link rel="stylesheet" media="all"
        href="../../_static/theme/css/default.css">
  <link rel="stylesheet" media="all"
        href="../../_static/theme/css/hieroglyph.css">
  <link rel="stylesheet" media="only screen and (max-device-width: 480px)"
        href="../../_static/theme/css/phone.css">

    

    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->
  
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>

    <script data-main="../../_static/js/slides"
            src="../../_static/js/require-1.0.8.min.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../../index.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


</head>
<body style="opacity: 0">

<slides class="layout-widescreen">

  

  
    <slide class="title-slide segue nobackground level-1" id="problem-set-3-key">
    <hgroup>
      <h1>Problem Set 3 : KEY<a class="headerlink" href="../../../Problem_Sets/problem_set_keys/problem-set-3-key.html#problem-set-3-key" title="View HTML">§</a></h1>
    </hgroup>
    <article class="">
      <p>Here is bash script that runs the analysis:</p>
<div class="highlight-bash"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73</pre></div></td><td class="code"><div class="highlight"><pre><span class="c">#! /usr/bin/env bash</span>

<span class="c">#BSUB -J ps5</span>
<span class="c">#BSUB -o ps5.%J.out</span>
<span class="c">#BSUB -e ps5.%J.err</span>

<span class="c"># run script for Problem Set 5 key</span>

<span class="nb">set</span> -o nounset -o pipefail -o errexit -x

<span class="c"># XXX this will be user-specific</span>
<span class="nv">project</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$HOME</span><span class="s2">/devel/bio-workshop/_test/problem-set-keys/problem-set-5&quot;</span>

<span class="c"># input files</span>
<span class="nv">data</span><span class="o">=</span><span class="s2">&quot;/vol1/opt/data&quot;</span>
<span class="nv">fasta</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$data</span><span class="s2">/hg19.fa&quot;</span>
<span class="nv">dnase_bed</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$data</span><span class="s2">/wgEncodeRegDnaseClusteredV2.bed.gz&quot;</span>
<span class="nv">ctcf_bed</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$data</span><span class="s2">/wgEncodeUwTfbsHelas3CtcfStdPkRep1.narrowPeak.gz&quot;</span>
<span class="nv">clustered_tfbs_bed</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$data</span><span class="s2">/wgEncodeRegTfbsClusteredV3.bed.gz&quot;</span>

<span class="nv">result</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$project</span><span class="s2">/results/2014-03-03&quot;</span>

<span class="k">if</span> <span class="o">[</span> ! -d <span class="nv">$result</span> <span class="o">]</span><span class="p">;</span> <span class="k">then</span>
    mkdir -p <span class="nv">$result</span>
<span class="k">fi</span>

<span class="c"># output files</span>
<span class="nv">ctcf_fasta</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/ctcf.chr22.fa&quot;</span>
<span class="nv">ctcf_chr22_bed</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/ctcf.chr22.bed&quot;</span>
<span class="nv">tfbs_no_dhs</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/tfbs_no_dhs.bed&quot;</span>
<span class="nv">tfbs_no_dhs_summary</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/tfbs_no_dhs.summary.tab&quot;</span>
<span class="nv">dhs_no_tfbs</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/dhs_no_tfbs.bed&quot;</span>
<span class="nv">dhs_no_tfbs_chr22</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/dhs_no_tfbs.chr22.bed&quot;</span>
<span class="nv">dhs_no_tfbs_fasta</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/dhs_no_tfbs.chr22.fa&quot;</span>
<span class="nv">problem_1_motifs</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/problem_1_motifs&quot;</span>
<span class="nv">problem_2_motifs</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$result</span><span class="s2">/problem_2_motifs&quot;</span>

<span class="c"># Problem 1.1, 1.2</span>
zcat <span class="nv">$ctcf_bed</span> <span class="p">|</span> awk <span class="s1">&#39;$1 == &quot;chr22&quot;&#39;</span> &gt; <span class="nv">$ctcf_chr22_bed</span>
bedtools getfasta -fi <span class="nv">$fasta</span> -bed <span class="nv">$ctcf_chr22_bed</span> -fo <span class="nv">$ctcf_fasta</span>

<span class="k">if</span> <span class="o">[[</span> ! -d <span class="nv">$problem_1_motifs</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then</span>
    meme -nmotifs <span class="m">5</span> -dna -minw <span class="m">6</span> -maxw <span class="m">20</span> -maxsize <span class="m">100000000</span> <span class="se">\</span>
        -o <span class="nv">$problem_1_motifs</span> <span class="nv">$ctcf_fasta</span>
<span class="k">fi</span>

<span class="c"># Problem 2</span>
<span class="c"># Identify transcription factor binding peaks that do not overlap with</span>
<span class="c"># DNase I hypersensitive sites</span>

bedtools intersect -a <span class="nv">$clustered_tfbs_bed</span> -b <span class="nv">$dnase_bed</span> -v -sorted <span class="se">\</span>
    &gt; <span class="nv">$tfbs_no_dhs</span>

<span class="c"># What transcription factors are represented in these peaks?</span>
cut -f4 <span class="nv">$tfbs_no_dhs</span> <span class="se">\</span>
    <span class="p">|</span> sort <span class="se">\</span>
    <span class="p">|</span> uniq -c <span class="se">\</span>
    <span class="p">|</span> sort -k1nr <span class="se">\</span>
    &gt; <span class="nv">$tfbs_no_dhs_summary</span>

<span class="c"># identify DNase I hypersensitive sites that do not have corresponding</span>
<span class="c"># transcription factor peak calls</span>
bedtools intersect -a <span class="nv">$dnase_bed</span> -b <span class="nv">$clustered_tfbs_bed</span> -v -sorted <span class="se">\</span>
    &gt; <span class="nv">$dhs_no_tfbs</span>

<span class="c"># What motifs are enriched in this set of hypersensitive sites?</span>
awk <span class="s1">&#39;$1 == &quot;chrY&quot; &amp;&amp; ($3 - $2 &gt; 8)&#39;</span> &lt; <span class="nv">$dhs_no_tfbs</span> &gt; <span class="nv">$dhs_no_tfbs_chr22</span>
bedtools getfasta -fi <span class="nv">$fasta</span> -bed <span class="nv">$dhs_no_tfbs_chr22</span> -fo <span class="nv">$dhs_no_tfbs_fasta</span>

<span class="k">if</span> <span class="o">[[</span> ! -d <span class="nv">$problem_2_motifs</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then</span>
    meme -nmotifs <span class="m">10</span> -dna -minw <span class="m">6</span> -maxw <span class="m">20</span> -maxsize <span class="m">100000000</span> <span class="se">\</span>
        -o <span class="nv">$problem_2_motifs</span> <span class="nv">$dhs_no_tfbs_fasta</span>
<span class="k">fi</span>
</pre></div>
</td></tr></table></div>




    </article>
  </slide>


    <slide class="thank-you-slide segue nobackground">
    <article class="flexbox vleft auto-fadein">
      <h2>&lt;Thank You!&gt;</h2>
    </article>
    <p class="auto-fadein" data-config-contact>
      <!-- populated from slide_config.json -->
    </p>
  </slide>

  <slide class="backdrop"></slide>

</slides>

<!--[if IE]>
  <script src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js"></script>
  <script>CFInstall.check({mode: 'overlay'});</script>
<![endif]-->
</body>
</html>