<!--
Google IO 2012/2013 HTML5 Slide Template

Authors: Eric Bidelman <ebidel@gmail.com>
         Luke Mahé <lukem@google.com>

URL: https://code.google.com/p/io-2012-slides
--><!DOCTYPE html>


<html>
<head>
  <title>Class 6 : Genomic Data Vignette: ChIP-seq &mdash; Genome Analysis Workshop 0.2 documentation</title>
  <meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">-->
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0">-->
  <!--This one seems to work all the time, but really small on ipad-->
  <!--<meta name="viewport" content="initial-scale=0.4">-->
  <meta name="apple-mobile-web-app-capable" content="yes">
  
  <link rel="stylesheet" media="all"
        href="../_static/theme/css/default.css">
  <link rel="stylesheet" media="all"
        href="../_static/theme/css/hieroglyph.css">
  <link rel="stylesheet" media="only screen and (max-device-width: 480px)"
        href="../_static/theme/css/phone.css">

    

    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->
  
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>

    <script data-main="../_static/js/slides"
            src="../_static/js/require-1.0.8.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Class 7 : BEDTools vignette" href="class-07.html" />
    <link rel="prev" title="Class 5 : BEDTools" href="class-05.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


</head>
<body style="opacity: 0">

<slides class="layout-widescreen">

    <slide class="title-slide segue nobackground">
    <aside class="gdbar"><img src="images/google_developers_icon_128.png"></aside>
    <!-- The content of this hgroup is replaced programmatically through the slide_config.json. -->
    <hgroup class="auto-fadein">
      <h1 data-config-title><!-- populated from slide_config.json --></h1>
      <h2 data-config-subtitle><!-- populated from slide_config.json --></h2>
      <p data-config-presenter><!-- populated from slide_config.json --></p>
    </hgroup>
  </slide>

  
    <slide class="level-1">
    <hgroup>
      <h1>Class 6 : Genomic Data Vignette: ChIP-seq<a class="headerlink" href="../../Classes/class-06.html#class-6-genomic-data-vignette-chip-seq" title="View HTML">§</a></h1>
    </hgroup>
    <article class="">
      <table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">2015 Feb 12 Thurs</td>
</tr>
<tr class="field-even field"><th class="field-name">Last updated:</th><td class="field-body">March 11, 2015</td>
</tr>
</tbody>
</table>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Goals<a class="headerlink" href="../../Classes/class-06.html#goals" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <ol class="arabic simple">
<li>Learn workflows for analyzing ChIP-seq data</li>
<li>Derive DNA sequence motifs from ChIP-seq peaks</li>
</ol>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>ENCODE<a class="headerlink" href="../../Classes/class-06.html#encode" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>The Human Genome Project was finished, giving us a list of human genes and their
locations. Unfortunately, we still had no idea how they were regulated. If only
there was an <a class="reference external" href="http://www.sciencemag.org.hsl-ezproxy.ucdenver.edu/content/306/5696/636.full">ENCyclopedia Of Dna Elements</a>…</p>
<p>Advantages: massive amounts of information on key cell lines, reproducible
experiments, public data access, technology development.</p>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>ENCODE Project Cell Lines<a class="headerlink" href="../../Classes/class-06.html#encode-project-cell-lines" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>Tier 1: GM12878 (EBV-transformed lymphoblast), K562 (CML lymphoblast), H1-hESC</p>
<p>Tier 2: HeLa-S3 (cervical cancer), HepG2 (liver carcinoma), HUVEC (umbilical vein)</p>
<p>Tier 2.5: SKNSH (neuroblastoma), IMR90 (lung fibroblast), A549 (lung carcinoma),
MCF7 (breast carcinoma), LHCN (myoblast), CD14+, CD20+</p>
<p><a class="reference external" href="http://genome.ucsc.edu/ENCODE/cellTypes.html">link</a> (this page also has very useful
links to cell culture protocols)</p>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Experiments<a class="headerlink" href="../../Classes/class-06.html#experiments" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <ol class="arabic simple">
<li>ChIP-seq: Histone marks, transcription factors</li>
<li>Chromatin structure: DNaseI-seq, FAIRE, 5C/Hi-C</li>
<li>RNA expression: mRNA-seq, GENCODE gene predictions</li>
<li>Data Integration: Segway / ChromHMM integration of functional data</li>
</ol>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Common File Formats<a class="headerlink" href="../../Classes/class-06.html#common-file-formats" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <ul class="simple">
<li>FASTQ: Raw sequencing data. <cite>[link] &lt;http://maq.sourceforge.net/fastq.shtml&gt;</cite></li>
<li>SAM/BAM: Aligned sequence data <cite>[link] &lt;http://samtools.github.io/hts-specs/SAMv1.pdf&gt;</cite></li>
<li>Bed/bigBed: List of genomic regions <cite>[link] &lt;http://genome.ucsc.edu/FAQ/FAQformat.html#format1&gt;</cite></li>
<li>Bedgraph/Wig/bigWig: Continuous signal <cite>[link] &lt;http://genome.ucsc.edu/goldenPath/help/bedgraph.html&gt;</cite></li>
</ul>
<p>Many other formats are described on this <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html">page</a></p>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>References<a class="headerlink" href="../../Classes/class-06.html#references" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>Completion of the entire project, and a ton of papers:
<a class="reference external" href="http://www.nature.com/nature/journal/v489/n7414/index.html">Nature</a>,
<a class="reference external" href="http://genome.cshlp.org/content/22/9.toc">Genome Research</a>,
<a class="reference external" href="http://genomebiology.com/content/13/9">Genome Biology</a>,</p>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>How to Access ENCODE Data<a class="headerlink" href="../../Classes/class-06.html#how-to-access-encode-data" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>The <a class="reference external" href="https://www.encodeproject.org/">ENCODE project page</a> is the portal
to all of the ENCODE data.</p>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Chromatin Immunoprecipitation Overview<a class="headerlink" href="../../Classes/class-06.html#chromatin-immunoprecipitation-overview" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>Chromatin Immunoprecipitation is used to determine where a protein of
interest binds on a chromatin template <a class="reference internal" href="#park-chipseq" id="id1">[Park_Chipseq]</a>.</p>
<table class="docutils citation" frame="void" id="park-chipseq" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[Park_Chipseq]</a></td><td><a class="reference external" href="http://www.nature.com/nrg/journal/v10/n10/full/nrg2641.html">http://www.nature.com/nrg/journal/v10/n10/full/nrg2641.html</a></td></tr>
</tbody>
</table>
<img alt="../_images/chip-workflow.png" src="../_images/chip-workflow.png" />




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Chromatin Immunoprecipitation Overview<a class="headerlink" href="../../Classes/class-06.html#id6" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <img alt="../_images/chip-data.png" src="../_images/chip-data.png" />




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>ChIP-seq analysis workflow<a class="headerlink" href="../../Classes/class-06.html#chip-seq-analysis-workflow" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>A general workflow for visualizing ChIP-seq data (and many other types of
data) is:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Operation</th>
<th class="head">File formats</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Align reads to reference genome</td>
<td><code class="docutils literal"><span class="pre">FASTQ</span> <span class="pre">~~&gt;</span> <span class="pre">BAM</span></code></td>
</tr>
<tr class="row-odd"><td>Generate coverage plots</td>
<td><code class="docutils literal"><span class="pre">BAM</span> <span class="pre">~~&gt;</span> <span class="pre">bedGraph</span></code></td>
</tr>
<tr class="row-even"><td>Call peaks</td>
<td><code class="docutils literal"><span class="pre">BAM</span> <span class="pre">~~&gt;</span> <span class="pre">BED</span></code></td>
</tr>
<tr class="row-odd"><td>Make binary files for UCSC display</td>
<td><code class="docutils literal"><span class="pre">bedGraph</span> <span class="pre">~~&gt;</span> <span class="pre">bigWig</span></code>, <code class="docutils literal"><span class="pre">BED</span> <span class="pre">~~&gt;</span> <span class="pre">bigBed</span></code></td>
</tr>
<tr class="row-even"><td>Identify motifs</td>
<td><code class="docutils literal"><span class="pre">BED</span> <span class="pre">~~&gt;</span> <span class="pre">FASTA</span> <span class="pre">~~&gt;</span> <span class="pre">TXT</span> <span class="pre">/</span> <span class="pre">HTML</span></code></td>
</tr>
</tbody>
</table>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>ChIP-seq data<a class="headerlink" href="../../Classes/class-06.html#chip-seq-data" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>Look at some human ChIP-seq data <a class="footnote-reference" href="#id3" id="id2">[1]</a>.</p>
<table class="docutils footnote" frame="void" id="id3" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[1]</a></td><td>Genome Browser Session
<a class="reference external" href="http://goo.gl/WfJxcM">http://goo.gl/WfJxcM</a></td></tr>
</tbody>
</table>
<p>(We'll talk more in depth about ChIP-Seq workflows in the future,
but for now, just a brief introdcution to a few commands you can use to
work on ChIP-Seq data (and pset3)).</p>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Peak calling<a class="headerlink" href="../../Classes/class-06.html#peak-calling" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>There are several available software packages for identifying regions
enriched in your IP experiment (i.e. peaks). We will use macs2 here.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># minimal macs2 command</span>
<span class="nv">$ </span>macs2 callpeak --treatment &lt;aln.bam&gt; --name &lt;exp.name&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>




    </article>
  </slide>  <slide class="level-2">
    <hgroup>
      <h2>Identify sequence motifs in enriched regions<a class="headerlink" href="../../Classes/class-06.html#identify-sequence-motifs-in-enriched-regions" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>You can use meme <a class="footnote-reference" href="#id5" id="id4">[2]</a> to identify over-represented motifs in groups of
sequences (e.g. sequences covered by ChIP peaks).</p>
<p>Use the <a class="reference external" href="http://bedtools.readthedocs.org/en/latest/content/tools/getfasta.html#getfasta" title="(in bedtools v2.22.1)"><span class="xref std std-ref">bedtools getfasta</span></a> command to fetch
fasta sequences.</p>
<p>Note: meme looks at both strands of a DNA sequence by default.</p>
<table class="docutils footnote" frame="void" id="id5" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id4">[2]</a></td><td>MEME
<a class="reference external" href="http://meme.nbcr.net/meme/">http://meme.nbcr.net/meme/</a></td></tr>
</tbody>
</table>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bedtools getfasta -fi &lt;ref.fa&gt; -bed &lt;peaks.bed&gt; -fo peaks.fa
<span class="nv">$ </span>meme -nmotifs <span class="m">5</span> -minw <span class="m">6</span> -maxw <span class="m">20</span> -dna &lt;peaks.fa&gt;
</pre></div>
</div>




    </article>
  </slide>


    <slide class="thank-you-slide segue nobackground">
    <aside class="gdbar right"><img src="images/google_developers_icon_128.png"></aside>
    <article class="flexbox vleft auto-fadein">
      <h2>&lt;Thank You!&gt;</h2>
    </article>
    <p class="auto-fadein" data-config-contact>
      <!-- populated from slide_config.json -->
    </p>
  </slide>

  <slide class="backdrop"></slide>

</slides>

<!--[if IE]>
  <script src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js"></script>
  <script>CFInstall.check({mode: 'overlay'});</script>
<![endif]-->
</body>
</html>