<!--
Google IO 2012/2013 HTML5 Slide Template

Authors: Eric Bidelman <ebidel@gmail.com>
         Luke Mahé <lukem@google.com>

URL: https://code.google.com/p/io-2012-slides
--><!DOCTYPE html>


<html>
<head>
  <title>Class 9 : R : Manipulation &amp; Plotting &mdash; Genome Analysis Workshop 0.2 documentation</title>
  <meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">-->
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0">-->
  <!--This one seems to work all the time, but really small on ipad-->
  <!--<meta name="viewport" content="initial-scale=0.4">-->
  <meta name="apple-mobile-web-app-capable" content="yes">

  <meta name="hieroglyph-title" data-config-title>
  <meta name="hieroglyph-subtitle" data-config-subtitle>
  <meta name="hieroglyph-presenter" data-config-presenter>

  
  <link rel="stylesheet" media="all"
        href="../_static/theme/css/default.css">
  <link rel="stylesheet" media="all"
        href="../_static/theme/css/hieroglyph.css">
  <link rel="stylesheet" media="only screen and (max-device-width: 480px)"
        href="../_static/theme/css/phone.css">

    

    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->
  
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>

    <script data-main="../_static/js/slides"
            src="../_static/js/require-1.0.8.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Class list" href="../Classes.html" />
    <link rel="next" title="Class 15 : Python : Basics" href="class-15.html" />
    <link rel="prev" title="Class 8 : R : Getting Started" href="class-08.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


</head>
<body style="opacity: 0">

<slides class="layout-widescreen">

  

  
    <slide class="title-slide segue nobackground level-1" id="class-9-r-manipulation-plotting">
    <hgroup>
      <h1>Class 9 : R : Manipulation &amp; Plotting<a class="headerlink" href="../../Classes/class-09.html#class-9-r-manipulation-plotting" title="View HTML">§</a></h1>
    </hgroup>
    <article class="">
      <table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Class date:</th><td class="field-body">2017 Feb 23 Thurs</td>
</tr>
<tr class="field-even field"><th class="field-name">Last updated:</th><td class="field-body">Mar 30, 2017</td>
</tr>
</tbody>
</table>




    </article>
  </slide>  <slide class="level-2" id="goals">
    <hgroup>
      <h2>Goals<a class="headerlink" href="../../Classes/class-09.html#goals" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <ol class="arabic simple">
<li>Begin to make plots with ggplot2 and combine with dplyr</li>
</ol>




    </article>
  </slide>  <slide class="level-2" id="ggplot2-syntax">
    <hgroup>
      <h2>ggplot2 syntax<a class="headerlink" href="../../Classes/class-09.html#ggplot2-syntax" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <div class="highlight-r"><div class="highlight"><pre><span></span><span class="kn">library</span><span class="p">(</span>tidyverse<span class="p">)</span>
df <span class="o">&lt;-</span> read_tsv<span class="p">(</span><span class="s">&#39;expr-geno-covs.txt&#39;</span><span class="p">)</span>

ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span>

ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
    geom_point<span class="p">()</span>

ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
    geom_point<span class="p">(</span>color<span class="o">=</span><span class="s">&#39;red&#39;</span><span class="p">,</span> size <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</pre></div>
</div>




    </article>
  </slide>  <slide class="level-2" id="id2">
    <hgroup>
      <h2>ggplot2 syntax (2)<a class="headerlink" href="../../Classes/class-09.html#id2" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p><code class="docutils literal"><span class="pre">aes()</span></code> stands for <strong>aesthetics</strong>, which specifies the the coordinates,
colors, size, etc from these columns in the data.frame.</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">,</span> color<span class="o">=</span>gender<span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">geom_point()</span></code> means plot these as points, could be <code class="docutils literal"><span class="pre">geom_line()</span></code> or
a number of other <cite>geoms</cite>.</p>




    </article>
  </slide>  <slide class="level-2" id="ggplot2-docs">
    <hgroup>
      <h2>ggplot2 docs<a class="headerlink" href="../../Classes/class-09.html#ggplot2-docs" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>The ggplot2 docs are very good: <a class="reference external" href="http://docs.ggplot2.org/current/">http://docs.ggplot2.org/current/</a></p>
<p>Use them find how to change the y-scale on this plot to log10:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>dfx<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
        geom_point<span class="p">()</span>
</pre></div>
</div>
<p>Look at the <cite>geom_point()</cite> documentation and change the color
of the plot above so that males and females are colored differently.</p>




    </article>
  </slide>  <slide class="level-2" id="histograms">
    <hgroup>
      <h2>Histograms<a class="headerlink" href="../../Classes/class-09.html#histograms" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>One of the simplest things to do in R, without ggplot is to look at
a histogram of your data:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>hist<span class="p">(</span>df<span class="o">$</span><span class="kp">expression</span><span class="p">)</span>
<span class="c1"># or</span>
hist<span class="p">(</span><span class="kp">log</span><span class="p">(</span>df<span class="o">$</span><span class="kp">expression</span><span class="p">))</span>
</pre></div>
</div>
<p>You can make these look a lot nicer with ggplot2.</p>
<p><code class="docutils literal"><span class="pre">hist()</span></code> does not work with ggplot, you'll have to use the
ggplot2 machinery for that:</p>




    </article>
  </slide>  <slide class="level-2" id="id3">
    <hgroup>
      <h2>Histograms (2)<a class="headerlink" href="../../Classes/class-09.html#id3" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span><span class="kp">log10</span><span class="p">(</span><span class="kp">expression</span><span class="p">))</span>
  <span class="o">+</span> geom_histogram<span class="p">()</span>
</pre></div>
</div>
<p>Try to figure out how to overlay a density plot with <code class="docutils literal"><span class="pre">geom_density()</span></code>.</p>




    </article>
  </slide>  <slide class="level-2" id="id4">
    <hgroup>
      <h2>Histograms (3)<a class="headerlink" href="../../Classes/class-09.html#id4" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>You can also make effective separate visualizations with <code class="docutils literal"><span class="pre">facet_grid()</span></code>
and <code class="docutils literal"><span class="pre">facet_wrap()</span></code>. You need to specify a formula to determine how to
separate:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>gp <span class="o">&lt;-</span> ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span> geom_point<span class="p">()</span>

<span class="c1"># separate by gender</span>
gp <span class="o">+</span> facet_grid<span class="p">(</span><span class="o">~</span> gender<span class="p">)</span>

<span class="c1"># separate by gender and condition</span>
gp <span class="o">+</span> facet_grid<span class="p">(</span>condition <span class="o">~</span> gender<span class="p">)</span>
</pre></div>
</div>




    </article>
  </slide>  <slide class="level-2" id="exercises">
    <hgroup>
      <h2>Exercises<a class="headerlink" href="../../Classes/class-09.html#exercises" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <ol class="arabic simple">
<li>Make a histogram using ggplot and separate cases from controls by
changing fill color and symbol type.</li>
<li>Make a histogram using ggplot and separate cases from controls
using <code class="docutils literal"><span class="pre">facet_grid()</span></code>, <code class="docutils literal"><span class="pre">facet_wrap()</span></code>.</li>
</ol>




    </article>
  </slide>  <slide class="level-3" id="exercises-with-dplyr">
    <hgroup>
      <h3>Exercises with dplyr<a class="headerlink" href="../../Classes/class-09.html#exercises-with-dplyr" title="View HTML">§</a></h3>
    </hgroup>
    <article class="">
      <p>Use the <code class="docutils literal"><span class="pre">expr-geno-covs.txt</span></code> file for the following exercises.</p>
<ol class="arabic simple">
<li>Use <code class="docutils literal"><span class="pre">dplyr</span></code> to calculate the mean age of smokers grouped by gender
and smoking status. Plot the result.</li>
<li>Make a plot of age by expression faceted by genotype. Fit a linear
model through these curves (use geom_smooth) on the plot.</li>
</ol>




    </article>
  </slide>  <slide class="level-2" id="combining-with-ggplot">
    <hgroup>
      <h2>Combining with ggplot<a class="headerlink" href="../../Classes/class-09.html#combining-with-ggplot" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>We had mean expression by condition and genotype as:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>df <span class="o">%&gt;%</span> group_by<span class="p">(</span>condition<span class="p">,</span> genotype<span class="p">)</span> \
    <span class="o">%&gt;%</span> summarize<span class="p">(</span>count <span class="o">=</span> n<span class="p">(),</span>
                  expr.mean <span class="o">=</span> <span class="kp">mean</span><span class="p">(</span><span class="kp">expression</span><span class="p">))</span>
</pre></div>
</div>
<p>We can continue the thought by adding a plot with the pipe:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span><span class="o">%&gt;%</span> ggplot<span class="p">(</span>aes<span class="p">(</span>x<span class="o">=</span>genotype<span class="p">,</span> y<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> \
    <span class="o">+</span> geom_histogram<span class="p">(</span>stat<span class="o">=</span><span class="s">&#39;identity&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>how can we change the color of all the bars to 'red'? [Hint, it's not
<strong>color</strong> ='red']</p>




    </article>
  </slide>  <slide class="level-2" id="ggplot-histograms">
    <hgroup>
      <h2>ggplot histograms<a class="headerlink" href="../../Classes/class-09.html#ggplot-histograms" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>Since <cite>expr-geno-covs.txt</cite> is already in long format, we can use it directly in
ggplot:</p>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
       geom_histogram<span class="p">()</span> <span class="o">+</span>
       scale_x_log10<span class="p">()</span>
</pre></div>
</div>




    </article>
  </slide>  <slide class="level-3" id="id1">
    <hgroup>
      <h3>Exercises<a class="headerlink" href="../../Classes/class-09.html#id1" title="View HTML">§</a></h3>
    </hgroup>
    <article class="">
      <ol class="arabic simple">
<li>Adjust this:</li>
</ol>
<div class="highlight-r"><div class="highlight"><pre><span></span>ggplot<span class="p">(</span>df<span class="p">,</span> aes<span class="p">(</span>x<span class="o">=</span><span class="kp">expression</span><span class="p">))</span> <span class="o">+</span>
       geom_histogram<span class="p">()</span> <span class="o">+</span>
       scale_x_log10<span class="p">()</span>
</pre></div>
</div>
<ul class="simple">
<li>to color by genotype</li>
<li>and to split plots (facet_wrap) by condition (case/control)</li>
<li>to color by age &gt; 60 vs. &lt;= 60 (use row selection stuff from start of class to
make a new column named, e.g. <cite>is_old</cite>)</li>
</ul>
<ol class="arabic simple">
<li>Figure out how to move overlapping points so categorical data is
viewable (hint: look at geom_jitter() or the <cite>position</cite> argument to
geom_point())</li>
<li>Load a BED file (e.g. <code class="docutils literal"><span class="pre">lamina.bed</span></code>) and calculate the mean length of
regions on each chromosome in the BED file with dplyr.  Plot the result as
a bar plot with ggplot2.</li>
</ol>




    </article>
  </slide>  <slide class="level-3" id="practice">
    <hgroup>
      <h3>Practice<a class="headerlink" href="../../Classes/class-09.html#practice" title="View HTML">§</a></h3>
    </hgroup>
    <article class="">
      <p>Start the Datacamp dplyr and ggplot2 courses if you haven't already.</p>




    </article>
  </slide>


    <slide class="thank-you-slide segue nobackground">
    <article class="flexbox vleft auto-fadein">
      <h2>&lt;Thank You!&gt;</h2>
    </article>
    <p class="auto-fadein" data-config-contact>
      <!-- populated from slide_config.json -->
    </p>
  </slide>

  <slide class="backdrop"></slide>

</slides>

<!--[if IE]>
  <script src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js"></script>
  <script>CFInstall.check({mode: 'overlay'});</script>
<![endif]-->
</body>
</html>