<!--
Google IO 2012/2013 HTML5 Slide Template

Authors: Eric Bidelman <ebidel@gmail.com>
         Luke Mahé <lukem@google.com>

URL: https://code.google.com/p/io-2012-slides
--><!DOCTYPE html>


<html>
<head>
  <title>FASTQ/A Exercises &mdash; Genome Analysis Workshop 0.2 documentation</title>
  <meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">-->
  <!--<meta name="viewport" content="width=device-width, initial-scale=1.0">-->
  <!--This one seems to work all the time, but really small on ipad-->
  <!--<meta name="viewport" content="initial-scale=0.4">-->
  <meta name="apple-mobile-web-app-capable" content="yes">

  <meta name="hieroglyph-title" data-config-title>
  <meta name="hieroglyph-subtitle" data-config-subtitle>
  <meta name="hieroglyph-presenter" data-config-presenter>

  
  <link rel="stylesheet" media="all"
        href="../_static/theme/css/default.css">
  <link rel="stylesheet" media="all"
        href="../_static/theme/css/hieroglyph.css">
  <link rel="stylesheet" media="only screen and (max-device-width: 480px)"
        href="../_static/theme/css/phone.css">

    

    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->
  
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>

    <script data-main="../_static/js/slides"
            src="../_static/js/require-1.0.8.min.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="top" title="Genome Analysis Workshop 0.2 documentation" href="../index.html" />
    <link rel="up" title="Exercises" href="../Exercises.html" />
    <link rel="next" title="Problem Sets" href="../Problem_Sets.html" />
    <link rel="prev" title="Exercises" href="exercises.html" /> 

<!-- Google Analytics tracking code -->

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-2', 'auto');
  ga('send', 'pageview');

</script>


</head>
<body style="opacity: 0">

<slides class="layout-widescreen">

  

  
    <slide class="title-slide segue nobackground level-1" id="fastq-a-exercises">
    <hgroup>
      <h1>FASTQ/A Exercises<a class="headerlink" href="../../Exercises/fastq_exercises.html#fastq-a-exercises" title="View HTML">§</a></h1>
    </hgroup>
    <article class="">
      <table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Last updated:</th><td class="field-body">Jan 26, 2017</td>
</tr>
</tbody>
</table>
<p>Exercises using unix tools to manipulate FASTQ/A files</p>




    </article>
  </slide>  <slide class="level-2" id="fastq-exercises">
    <hgroup>
      <h2>FASTQ exercises<a class="headerlink" href="../../Exercises/fastq_exercises.html#fastq-exercises" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>Download a fastq example file <a class="reference download internal" href="../_downloads/SP1.fq" download=""><code class="xref download docutils literal"><span class="pre">[SP1.fq]</span></code></a></p>
<p>The FASTQ file <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">format</a> contains 4 lines per read sequence.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ head -n <span class="m">4</span> SP1.fq
@cluster_2:UMI_ATTCCG
TTTCCGGGGCACATAATCTTCAGCCGGGCGC
+
9C<span class="p">;</span><span class="o">=</span><span class="p">;</span><span class="o">=</span>&lt;9@4868&gt;9:67AA&lt;9&gt;65&lt;<span class="o">=</span>&gt;591
</pre></div>
</div>
<ol class="arabic simple">
<li>Line 1 contains a sequence identifier that begins with an <cite>&#64;</cite></li>
<li>Line 2 contains the read sequence (A,T,C,G,N)</li>
<li>Line 3 is a comment line, often unused and only contains a <cite>+</cite></li>
<li>Line 4 is the Phred quality score for each sequence encoded in ASCII format</li>
</ol>
<p>First use <code class="docutils literal"><span class="pre">wc</span></code> and <code class="docutils literal"><span class="pre">awk</span></code> to determine the number of <em>sequences</em> in the fastq</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ wc -l SP1.fq  <span class="c1"># total number of lines</span>
$ wc -l SP1.fq <span class="p">|</span> awk <span class="s1">&#39;{print $1 / 4}&#39;</span> <span class="c1"># total number of fastq records</span>
</pre></div>
</div>
<p>A common mistake is to use <code class="docutils literal"><span class="pre">grep</span></code> to pattern match the <code class="docutils literal"><span class="pre">&#64;</span></code> in the
sequence identifier. Why doesn't this work?</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ wc -l SP1.fq <span class="p">|</span> awk <span class="s1">&#39;{print $1 / 4}&#39;</span>
250
$ grep -c <span class="s2">&quot;@&quot;</span> SP1.fq
459
</pre></div>
</div>
<p>Next, extract out the read <em>sequences</em> from the fastq. This is a bit
complicated as we need to only pull out the second line from each record.</p>
<p>One approach to this problem is to use the <code class="docutils literal"><span class="pre">%</span></code> <a class="reference external" href="https://en.wikipedia.org/wiki/Modulo_operation">modulo
operator</a>, which returns
the remainder after division of two integers. For example using <code class="docutils literal"><span class="pre">awk</span></code>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ awk <span class="s1">&#39;BEGIN { {print 4 % 2}}&#39;</span>
$ awk <span class="s1">&#39;BEGIN { {print 4 % 3}}&#39;</span>
$ awk <span class="s1">&#39;BEGIN { {print 5 % 3}}&#39;</span>
$ awk <span class="s1">&#39;BEGIN { {print 1 % 4}}&#39;</span>
</pre></div>
</div>
<p>In <code class="docutils literal"><span class="pre">awk</span></code> there is a special variable <code class="docutils literal"><span class="pre">NR</span></code> which is equal to the
current line number.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ awk <span class="s1">&#39;{print NR}&#39;</span> SP1.fq
</pre></div>
</div>
<p>We can use the modulo operator with <code class="docutils literal"><span class="pre">NR</span></code> to only capture specific
records from the fastq</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ awk <span class="s1">&#39;{print NR % 4, $0 }&#39;</span> SP1.fq <span class="p">|</span> head <span class="c1"># note output in first column</span>
<span class="m">1</span> @cluster_2:UMI_ATTCCG
<span class="m">2</span> TTTCCGGGGCACATAATCTTCAGCCGGGCGC
<span class="m">3</span> +
<span class="m">0</span> 9C<span class="p">;</span><span class="o">=</span><span class="p">;</span><span class="o">=</span>&lt;9@4868&gt;9:67AA&lt;9&gt;65&lt;<span class="o">=</span>&gt;591
<span class="m">1</span> @cluster_8:UMI_CTTTGA
<span class="m">2</span> TATCCTTGCAATACTCTCCGAACGGGAGAGC
<span class="m">3</span> +
<span class="m">0</span> 1/04.72,<span class="o">(</span>003,-2-22+00-12./.-.4-
<span class="m">1</span> @cluster_12:UMI_GGTCAA
<span class="m">2</span> GCAGTTTAAGATCATTTTATTGAAGAGCAAG


$ awk <span class="s1">&#39;NR % 4 == 1&#39;</span> SP1.fq <span class="p">|</span> head  <span class="c1"># get header line</span>
$ awk <span class="s1">&#39;NR % 4 == 2&#39;</span> SP1.fq <span class="p">|</span> head  <span class="c1"># get sequence line</span>
$ awk <span class="s1">&#39;NR % 4 == 3&#39;</span> SP1.fq <span class="p">|</span> head  <span class="c1"># get comment line</span>
$ awk <span class="s1">&#39;NR % 4 == 0&#39;</span> SP1.fq <span class="p">|</span> head  <span class="c1"># get quality line</span>
</pre></div>
</div>
<p>Now that we can isolate only the sequences let's use <cite>sort</cite> and <cite>uniq</cite> to
find common sequences.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>awk <span class="s1">&#39;NR % 4 == 2&#39;</span> SP1.fq <span class="se">\ </span> <span class="c1"># get sequences</span>
  <span class="p">|</span> sort <span class="se">\ </span> <span class="c1"># sort the sequences</span>
  <span class="p">|</span> uniq -c <span class="se">\ </span><span class="c1"># report number of occurances of each unique sequence</span>
  <span class="p">|</span> head

awk <span class="s1">&#39;NR % 4 == 2&#39;</span> SP1.fq <span class="se">\ </span> <span class="c1">#show most frequent sequences</span>
  <span class="p">|</span> sort <span class="se">\</span>
  <span class="p">|</span> uniq -c <span class="se">\</span>
  <span class="p">|</span> sort -k1,1nr <span class="se">\ </span><span class="c1"># reverse sort to rank by most common sequence</span>
  <span class="p">|</span> head
</pre></div>
</div>
<p>Next, using <code class="docutils literal"><span class="pre">cut</span></code>, <code class="docutils literal"><span class="pre">sort</span></code> and <code class="docutils literal"><span class="pre">uniq</span></code> lets look at the most common
sequences found in the first 10 base pairs of the read.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>awk <span class="s1">&#39;NR % 4 == 2&#39;</span> SP1.fq <span class="se">\ </span> <span class="c1"># get sequences</span>
<span class="p">|</span> cut -c 1-10 <span class="se">\ </span><span class="c1"># cut by character position 1-10</span>
<span class="p">|</span> sort <span class="se">\</span>
<span class="p">|</span> uniq - c <span class="se">\</span>
<span class="p">|</span> sort -k1,1nr <span class="se">\</span>
<span class="p">|</span> head
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">awk</span></code> can also be used directly to extract out sub-strings using the <code class="docutils literal"><span class="pre">substr()</span></code>
function</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="c1"># substr($field_index, start, end)</span>
$ awk <span class="s1">&#39;NR % 4 == 1 {print substr($0, 1, 10)}&#39;</span> SP1.fq <span class="se">\</span>
  <span class="p">|</span> head -5
</pre></div>
</div>
<p>Let's extract out the Unique Molecular Identifyier (UMI)
tag from the read identifier using <code class="docutils literal"><span class="pre">cut</span></code></p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ awk <span class="s1">&#39;NR % 4 == 1&#39;</span> SP1.fq <span class="p">|</span> head -5 <span class="c1"># get ids</span>
@cluster_2:UMI_ATTCCG
@cluster_8:UMI_CTTTGA
@cluster_12:UMI_GGTCAA
@cluster_21:UMI_AGAACA
@cluster_29:UMI_GCAGGA

$ awk <span class="s1">&#39;NR % 4 == 1&#39;</span> SP1.fq <span class="se">\</span>
  <span class="p">|</span> cut -d <span class="s2">&quot;:&quot;</span> -f <span class="m">2</span> <span class="se">\</span>
  <span class="p">|</span> head -5
</pre></div>
</div>
<p>Lastly, let's write a short awk program that uses the <code class="docutils literal"><span class="pre">substr()</span></code>
function to trim the first 10 bases off of the read and return the
sequences in FASTQ. Note that the length of the quality line needs to
match the length of the sequence line.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ awk <span class="s1">&#39;NR % 4 == 1 {print $0}; # print id line</span>
<span class="s1">  NR % 4 == 2 {print substr($0, 10, length($0))}; #extract from 10 to the end</span>
<span class="s1">  NR % 4 == 3 {print $0} ; # print comment line</span>
<span class="s1">  NR % 4 == 0 {print substr($0, 10, length($0))}&#39;</span> <span class="se">\ </span><span class="c1"># extract from pos 10 to the end</span>
  SP1.fq <span class="se">\</span>
  <span class="p">|</span> head

@cluster_2:UMI_ATTCCG
CACATAATCTTCAGCCGGGCGC
+
4868&gt;9:67AA&lt;9&gt;65&lt;<span class="o">=</span>&gt;591
@cluster_8:UMI_CTTTGA
AATACTCTCCGAACGGGAGAGC
+
003,-2-22+00-12./.-.4-
@cluster_12:UMI_GGTCAA
GATCATTTTATTGAAGAGCAAG
</pre></div>
</div>




    </article>
  </slide>  <slide class="level-2" id="fasta-exercises">
    <hgroup>
      <h2>FASTA exercises<a class="headerlink" href="../../Exercises/fastq_exercises.html#fasta-exercises" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <p>FASTA format is a more compact sequence format than FASTQ</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>&gt;sequence_identifier_1
ATCGTCGATGCTAGTCGA
&gt;sequence_identifier_2
AGCTAGCTAGCTAGC
</pre></div>
</div>
<p>Using just <code class="docutils literal"><span class="pre">awk</span></code> we can easily convert from FASTQ to FASTA</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ awk <span class="s1">&#39;NR % 4 == 1 {print &quot;&gt;&quot;$1};</span>
<span class="s1">  NR % 4 == 2 {print}&#39;</span> SP1.fq <span class="se">\</span>
  <span class="p">|</span> head

$ awk <span class="s1">&#39;NR % 4 == 1 {print &quot;&gt;&quot;substr($0, 2, length($0))};</span>
<span class="s1">  NR % 4 == 2 {print}&#39;</span> SP1.fq <span class="se">\</span>
  <span class="p">|</span> head  <span class="c1"># same as above but remove ugly @ character</span>

$ awk <span class="s1">&#39;NR % 4 == 1 {print &quot;&gt;&quot;substr($0, 2, length($0))};</span>
<span class="s1">  NR % 4 == 2 {print}&#39;</span> SP1.fq <span class="se">\</span>
  &gt; example.fa
</pre></div>
</div>
<p>To count the number of sequences in the fasta file you can use either
<code class="docutils literal"><span class="pre">wc</span></code> with <code class="docutils literal"><span class="pre">awk</span></code> as shown above or <code class="docutils literal"><span class="pre">grep</span></code> as follows:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>grep -c <span class="s2">&quot;&gt;&quot;</span> example.fa
</pre></div>
</div>




    </article>
  </slide>  <slide class="level-2" id="on-your-own">
    <hgroup>
      <h2>On your own<a class="headerlink" href="../../Exercises/fastq_exercises.html#on-your-own" title="View HTML">§</a></h2>
    </hgroup>
    <article class="">
      <ol class="arabic simple">
<li>Use <code class="docutils literal"><span class="pre">awk</span></code> to extract out sequences and find out the lengths of each
read. Are they all the same length?</li>
<li>Use <code class="docutils literal"><span class="pre">grep</span></code> and/or <code class="docutils literal"><span class="pre">awk</span></code> to find sequences with more than 4 A's in a row. How many
sequences did you find? Can you also extract out their names (hint see
man grep)? Do any UMI's have 4 A's in a row?</li>
<li>What is the most common hexamer found at the end of the reads?</li>
</ol>




    </article>
  </slide>


    <slide class="thank-you-slide segue nobackground">
    <article class="flexbox vleft auto-fadein">
      <h2>&lt;Thank You!&gt;</h2>
    </article>
    <p class="auto-fadein" data-config-contact>
      <!-- populated from slide_config.json -->
    </p>
  </slide>

  <slide class="backdrop"></slide>

</slides>

<!--[if IE]>
  <script src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js"></script>
  <script>CFInstall.check({mode: 'overlay'});</script>
<![endif]-->
</body>
</html>